<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>1. Differential expression</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="01_diffexp_sleuth_files/libs/clipboard/clipboard.min.js"></script>
<script src="01_diffexp_sleuth_files/libs/quarto-html/quarto.js"></script>
<script src="01_diffexp_sleuth_files/libs/quarto-html/popper.min.js"></script>
<script src="01_diffexp_sleuth_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="01_diffexp_sleuth_files/libs/quarto-html/anchor.min.js"></script>
<link href="01_diffexp_sleuth_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="01_diffexp_sleuth_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="01_diffexp_sleuth_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="01_diffexp_sleuth_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="01_diffexp_sleuth_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">1. Differential expression</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>V Chong | vanessa.chong@rdm.ox.ac.uk | 29 October 2025</p>
<section id="preamble" class="level1">
<h1>Preamble</h1>
<ul>
<li>Bulk RNAseq samples from healthy, patient-derived organoids of the terminal ileum (TI).</li>
<li>Libraries were prepared and sequenced in several batches.</li>
<li>Organoids were stimulated with IL22 only, both IL22 and IL26, or IL26 only. Controls were unstimulated.</li>
<li>Stimulation was performed for 24, 48, or 72 hours.</li>
</ul>
<section id="points-of-note" class="level2">
<h2 class="anchored" data-anchor-id="points-of-note">Points-of-note</h2>
<ul>
<li>24 hours samples are spread across three batches, and two different (previous and current) growth mediums. <strong>technical replicates in progress</strong></li>
<li>48 hours samples constitute a batch of its own, performed in current growth medium only. <strong>technical replicates in progress</strong></li>
<li>72 hours samples are not replicated and performed in a previous growth medium only.</li>
</ul>
<p>These are all reflected in the sample metadata called into the analysis later.</p>
</section>
</section>
<section id="analysis" class="level1">
<h1>Analysis</h1>
<p>Consider the impact of experimental covariates e.g.&nbsp;donor, batch, medium. Statistically test for genes that are affected by (any) IL treatment while controlling for differences due to these covariates. More detailed pairwise comparison to follow.</p>
<section id="pseudoalignment-and-estimation-of-transcript-abundances" class="level2">
<h2 class="anchored" data-anchor-id="pseudoalignment-and-estimation-of-transcript-abundances">Pseudoalignment and estimation of transcript abundances</h2>
<p><a href="https://doi.org/10.1038/nbt.3519">Kallisto</a> with bootstraps (<code>-b 100</code>) enabled for downstream <a href="https://doi.org/10.1038/nmeth.4324">sleuth</a> differential expression.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">module</span> load kallisto/0.46.1</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="ex">kallisto</span> quant <span class="at">-i</span> /databank/igenomes/Homo_sapiens/Ensembl/GRCh38/Sequence/Kallisto/cdna-release-85.idx <span class="at">-o</span> ./A2_11_A1 <span class="at">-b</span> 100 <span class="at">-t</span> 4 <span class="dt">\</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>/project/thelior_project/vchong/novogene_exp4_A2-11_A2-12/fastp/A2_11_A1.R1.fq.gz <span class="dt">\</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>/project/thelior_project/vchong/novogene_exp4_A2-11_A2-12/fastp/A2_11_A1.R2.fq.gz</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="differential-expression-analysis" class="level2">
<h2 class="anchored" data-anchor-id="differential-expression-analysis">Differential expression analysis</h2>
<p><a href="https://pachterlab.github.io/sleuth/walkthroughs">Reference</a> - “Differential analysis using p-value aggregation, on multiple conditions and covariates”</p>
<section id="initialise-r-environment" class="level3">
<h3 class="anchored" data-anchor-id="initialise-r-environment">Initialise R environment</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="do">##### Set libPaths #####</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">.libPaths</span>(<span class="fu">c</span>(<span class="st">"/home/chongmorrison/R/x86_64-pc-linux-gnu-library/4.5"</span>,</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>            <span class="st">"/home/chongmorrison/R/4.5.0-Bioc3.21"</span>))</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">.libPaths</span>() <span class="co"># check .libPaths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "/home/chongmorrison/R/x86_64-pc-linux-gnu-library/4.5"
[2] "/home/chongmorrison/R/4.5.0-Bioc3.21"                 
[3] "/usr/local/lib/R/site-library"                        
[4] "/usr/lib/R/site-library"                              
[5] "/usr/lib/R/library"                                   </code></pre>
</div>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="do">##### Memory/parallel cores usage #####</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">Ncpus =</span> <span class="dv">12</span>) <span class="co"># adjust no. of cores for base R</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="fu">getOption</span>(<span class="st">"Ncpus"</span>, <span class="dv">1</span>L)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 12</code></pre>
</div>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="do">##### Import all required packages #####</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="co"># data-wrangling and plots</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="co">#library(RColorBrewer)</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="co">#library(viridis)</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="co">#library(ggpubr)</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SummarizedExperiment)</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dittoSeq)</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a><span class="co"># analysis</span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(biomaRt)</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(PCAtools)</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(sleuth)</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rbioapi)</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a><span class="co"># to cite packages</span></span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(grateful)</span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a><span class="do">##### Set seed for reproducibility of random sampling #####</span></span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">584</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="ensembl-annotations---release-80" class="level3">
<h3 class="anchored" data-anchor-id="ensembl-annotations---release-80">Ensembl annotations - <strong>release 80</strong></h3>
<p><strong>Kallisto index on CCB was built on Ensembl release 85 (Sept 2016). Unfortunately the only currently-available archived versions of Ensembl from around this time are release 80 (May 2015) and release 100 (Apr 2020)</strong></p>
<p>Using release 80 annotations as it is the “closest” available and <strong>should not</strong> differ significantly to release 85.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># run once and save locally for later use</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>ensembl <span class="ot">&lt;-</span> biomaRt<span class="sc">::</span><span class="fu">useEnsembl</span>(<span class="at">biomart =</span> <span class="st">"genes"</span>, <span class="at">dataset =</span> <span class="st">"hsapiens_gene_ensembl"</span>, <span class="at">version =</span> <span class="dv">80</span>)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>annotations <span class="ot">&lt;-</span> biomaRt<span class="sc">::</span><span class="fu">getBM</span>(<span class="at">mart =</span> ensembl, <span class="at">attributes=</span><span class="fu">c</span>(<span class="st">"ensembl_transcript_id"</span>, <span class="st">"transcript_version"</span>, <span class="st">"external_gene_name"</span>, <span class="st">"ensembl_gene_id"</span>))</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>annotations<span class="sc">$</span>ensembl_transcript_id_version <span class="ot">&lt;-</span> <span class="fu">paste</span>(annotations<span class="sc">$</span>ensembl_transcript_id, annotations<span class="sc">$</span>transcript_version, <span class="at">sep=</span><span class="st">"."</span>)</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="do">## transcript-to-gene mapping</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>ttg <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">rename</span>(annotations, <span class="at">target_id =</span> ensembl_transcript_id_version, <span class="at">ens_gene =</span> ensembl_gene_id, <span class="at">ext_gene =</span> external_gene_name)</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>ttg <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">select</span>(ttg, <span class="fu">c</span>(<span class="st">'target_id'</span>, <span class="st">'ens_gene'</span>, <span class="st">'ext_gene'</span>))</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(ttg, <span class="at">file =</span> <span class="st">"./ttg85.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>ttg <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"./ttg85.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="load-sample-metadata" class="level3">
<h3 class="anchored" data-anchor-id="load-sample-metadata">Load sample metadata</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="at">file =</span> <span class="st">'./metadata_29.10.25.csv'</span>, <span class="at">sep =</span> <span class="st">','</span>, <span class="at">header=</span><span class="cn">TRUE</span>)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(metadata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      sample      donor      stim time batch medium
1    Ctrl_24 TIP-OX-924       ctl  24h   one      x
2    Ctrl_72 TIP-OX-924       ctl  72h   one      x
3    IL22_24 TIP-OX-924      IL22  24h   one      x
4 IL22_26_24 TIP-OX-924 IL22+IL26  24h   one      x
5 IL22_26_72 TIP-OX-924 IL22+IL26  72h   one      x
6    IL22_72 TIP-OX-924      IL22  72h   one      x
                                                                                                                           path
1    /home/chongmorrison/Dropbox/VCM-StarBook/ResearchProjects/theliorbio/analyses/rnaseq/kallisto_v0.46.1/Ctrl_24/abundance.h5
2    /home/chongmorrison/Dropbox/VCM-StarBook/ResearchProjects/theliorbio/analyses/rnaseq/kallisto_v0.46.1/Ctrl_72/abundance.h5
3    /home/chongmorrison/Dropbox/VCM-StarBook/ResearchProjects/theliorbio/analyses/rnaseq/kallisto_v0.46.1/IL22_24/abundance.h5
4 /home/chongmorrison/Dropbox/VCM-StarBook/ResearchProjects/theliorbio/analyses/rnaseq/kallisto_v0.46.1/IL22_26_24/abundance.h5
5 /home/chongmorrison/Dropbox/VCM-StarBook/ResearchProjects/theliorbio/analyses/rnaseq/kallisto_v0.46.1/IL22_26_72/abundance.h5
6    /home/chongmorrison/Dropbox/VCM-StarBook/ResearchProjects/theliorbio/analyses/rnaseq/kallisto_v0.46.1/IL22_72/abundance.h5</code></pre>
</div>
</div>
</section>
<section id="explore-experimental-covariates-vs-conditions" class="level3">
<h3 class="anchored" data-anchor-id="explore-experimental-covariates-vs-conditions">Explore experimental covariates vs conditions</h3>
<p>Build <strong>sleuth</strong> object from metadata and <strong>kallisto</strong> abundances.</p>
<p><code>target_mapping</code> = transcript groupings into genes, where <code>aggregation_column</code> is the column in <code>target_mapping</code> used to aggregate the transcripts. When both <code>target_mapping</code> and <code>aggregation_column</code> are provided, <strong>sleuth</strong> will automatically run in gene mode, <strong>returning gene differential expression results that came from the aggregation of transcript p-values</strong>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>so <span class="ot">&lt;-</span> <span class="fu">sleuth_prep</span>(metadata, </span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>                  <span class="at">target_mapping =</span> ttg, <span class="at">aggregation_column =</span> <span class="st">"ens_gene"</span>, </span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>                  <span class="at">extra_bootstrap_summary =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in check_num_cores(num_cores): It appears that you are running Sleuth from within Rstudio.
Because of concerns with forking processes from a GUI, 'num_cores' is being set to 1.
If you wish to take advantage of multiple cores, please consider running sleuth from the command line.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>reading in kallisto results</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>dropping unused factor levels</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>....................................
normalizing est_counts
65701 targets passed the filter
normalizing tpm
merging in metadata
summarizing bootstraps
....................................</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># covariates</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(so, <span class="at">units =</span> <span class="st">"tpm"</span>, <span class="at">color_by =</span> <span class="st">"donor"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: `aes_string()` was deprecated in ggplot2 3.0.0.
ℹ Please use tidy evaluation idioms with `aes()`.
ℹ See also `vignette("ggplot2-in-packages")` for more information.
ℹ The deprecated feature was likely used in the sleuth package.
  Please report the issue at &lt;https://github.com/pachterlab/sleuth/issues&gt;.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(so, <span class="at">units =</span> <span class="st">"tpm"</span>, <span class="at">color_by =</span> <span class="st">"batch"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-7-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(so, <span class="at">units =</span> <span class="st">"tpm"</span>, <span class="at">color_by =</span> <span class="st">"medium"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-7-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># conditions</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(so, <span class="at">units =</span> <span class="st">"tpm"</span>, <span class="at">color_by =</span> <span class="st">"time"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(so, <span class="at">units =</span> <span class="st">"tpm"</span>, <span class="at">color_by =</span> <span class="st">"stim"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-8-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<ul>
<li>72h samples should be excluded from the other samples.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>metadata <span class="ot">&lt;-</span> <span class="fu">subset</span>(metadata, <span class="sc">!</span>time <span class="sc">==</span> <span class="st">"72h"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>so <span class="ot">&lt;-</span> <span class="fu">sleuth_prep</span>(metadata, </span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>                  <span class="at">target_mapping =</span> ttg, <span class="at">aggregation_column =</span> <span class="st">"ens_gene"</span>, </span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>                  <span class="at">extra_bootstrap_summary =</span> <span class="cn">TRUE</span>,</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>                  <span class="at">num_cores =</span> <span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save sleuth object to reload without recomputing bootstraps</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(so, <span class="at">file =</span> <span class="st">"./01_diffexp_sleuth-so.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>so <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"./01_diffexp_sleuth-so.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># covariates</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(so, <span class="at">units =</span> <span class="st">"tpm"</span>, <span class="at">color_by =</span> <span class="st">"donor"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(so, <span class="at">units =</span> <span class="st">"tpm"</span>, <span class="at">color_by =</span> <span class="st">"batch"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-13-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(so, <span class="at">units =</span> <span class="st">"tpm"</span>, <span class="at">color_by =</span> <span class="st">"medium"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-13-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># conditions</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(so, <span class="at">units =</span> <span class="st">"tpm"</span>, <span class="at">color_by =</span> <span class="st">"time"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(so, <span class="at">units =</span> <span class="st">"tpm"</span>, <span class="at">color_by =</span> <span class="st">"stim"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-14-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="test-for-differentially-expressed-genes---multi-condition" class="level3">
<h3 class="anchored" data-anchor-id="test-for-differentially-expressed-genes---multi-condition">Test for differentially-expressed genes - multi-condition</h3>
<p>Fit a <code>reduced</code> model, which includes the parameters corresponding to experimental covariates we are controlling for i.e.&nbsp;<em>batch</em> (which will account for <em>medium</em>, <em>time</em>) and <em>donor</em>. At this point <em>time</em> is accounted for i.e.&nbsp;to control for genes that may be differentially expressed at different timepoints independently of <em>stim</em>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>so <span class="ot">&lt;-</span> <span class="fu">sleuth_fit</span>(so, <span class="sc">~</span>batch <span class="sc">+</span> donor, <span class="st">'reduced'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>fitting measurement error models</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>shrinkage estimation</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>1 NA values were found during variance shrinkage estimation due to mean observation values outside of the range used for the LOESS fit.
The LOESS fit will be repeated using exact computation of the fitted surface to extrapolate the missing values.
These are the target ids with NA values: ENST00000219022.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>computing variance of betas</code></pre>
</div>
</div>
<p>Fit the <code>full</code> model which includes all parameters including <em>stim</em>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>so <span class="ot">&lt;-</span> <span class="fu">sleuth_fit</span>(so, <span class="sc">~</span>batch <span class="sc">+</span> donor <span class="sc">+</span> stim, <span class="st">'full'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>fitting measurement error models</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>shrinkage estimation</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>3 NA values were found during variance shrinkage estimation due to mean observation values outside of the range used for the LOESS fit.
The LOESS fit will be repeated using exact computation of the fitted surface to extrapolate the missing values.
These are the target ids with NA values: ENST00000219022.2, ENST00000356495.4, ENST00000361624.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>computing variance of betas</code></pre>
</div>
</div>
<p>Then compare the <code>full</code> model to the <code>reduced</code> model with the likelihood ratio test. Using the likelihood ratio test, sleuth identifies genes whose abundances are significantly better explained when IL treatment (<em>stim</em>) is taken into account, while accounting for baseline differences that may be explained by <em>batch</em> or <em>donor</em>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>so <span class="ot">&lt;-</span> <span class="fu">sleuth_lrt</span>(so, <span class="st">'reduced'</span>, <span class="st">'full'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Obtain gene-level differential expression results. Sleuth uses the p-values from comparing transcripts to make a gene-level determination and perform gene differential expression.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>sleuth_table_gene <span class="ot">&lt;-</span> <span class="fu">sleuth_results</span>(so, <span class="st">'reduced:full'</span>, <span class="st">'lrt'</span>, <span class="at">show_all =</span> <span class="cn">FALSE</span>)</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>sleuth_table_gene <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(sleuth_table_gene, qval <span class="sc">&lt;=</span> <span class="fl">0.01</span>)</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sleuth_table_gene, <span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         target_id ext_gene num_aggregated_transcripts sum_mean_obs_counts
1  ENSG00000115386    REG1A                          4           13.747139
2  ENSG00000197249 SERPINA1                          5           17.529003
3  ENSG00000176153     GPX2                          5           31.790580
4  ENSG00000140279    DUOX2                          3           12.233783
5  ENSG00000179403     VWA1                          4           18.011017
6  ENSG00000140274   DUOXA2                          2           12.252341
7  ENSG00000165188   RNF183                          4           14.096104
8  ENSG00000115474   KCNJ13                          3           18.130287
9  ENSG00000253368    TRNP1                          2           13.655944
10 ENSG00000079385  CEACAM1                          4           17.818769
11 ENSG00000272398     CD24                          5           22.221984
12 ENSG00000163734    CXCL3                          3           11.822161
13 ENSG00000128268    MGAT3                          3           15.139323
14 ENSG00000185499     MUC1                          3           11.365316
15 ENSG00000205403      CFI                          2            5.898606
16 ENSG00000187908    DMBT1                          1            9.620689
17 ENSG00000176532    PRR15                          2           14.328065
18 ENSG00000095932   SMIM24                          3           18.655423
19 ENSG00000134827     TCN1                          2           13.827779
20 ENSG00000171236     LRG1                          2           10.986103
           pval         qval
1  2.128312e-69 2.620165e-65
2  1.651303e-45 1.016460e-41
3  3.608119e-44 1.480652e-40
4  5.263609e-38 1.620007e-34
5  8.142610e-37 2.004873e-33
6  1.812038e-36 3.718000e-33
7  1.084673e-35 1.907630e-32
8  1.823392e-29 2.805972e-26
9  3.505081e-28 4.794561e-25
10 3.145516e-25 3.872445e-22
11 6.291171e-25 7.040964e-22
12 8.496262e-25 8.716456e-22
13 4.908217e-24 4.229417e-21
14 4.752487e-24 4.229417e-21
15 5.153217e-24 4.229417e-21
16 1.460252e-23 1.123573e-20
17 1.955211e-23 1.415918e-20
18 4.173783e-23 2.854635e-20
19 5.559889e-23 3.602515e-20
20 2.592034e-22 1.595526e-19</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save result table</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(sleuth_table_gene, <span class="at">file =</span> <span class="st">"./outputs/01_diffexp_sleuth-genes-qval001-stim.csv"</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Obtain transcript-level differential expression results.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>sleuth_table_tx <span class="ot">&lt;-</span> <span class="fu">sleuth_results</span>(so, <span class="st">'reduced:full'</span>, <span class="st">'lrt'</span>, <span class="at">show_all =</span> <span class="cn">FALSE</span>, <span class="at">pval_aggregate =</span> <span class="cn">FALSE</span>)</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>sleuth_table_tx <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(sleuth_table_tx, qval <span class="sc">&lt;=</span> <span class="fl">0.01</span>)</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sleuth_table_tx, <span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           target_id        ens_gene ext_gene         pval         qval
1  ENST00000244565.7            &lt;NA&gt;     &lt;NA&gt; 2.035618e-24 6.534266e-20
2  ENST00000330163.8            &lt;NA&gt;     &lt;NA&gt; 3.057968e-24 6.534266e-20
3  ENST00000368956.6            &lt;NA&gt;     &lt;NA&gt; 3.057968e-24 6.534266e-20
4  ENST00000344338.7            &lt;NA&gt;     &lt;NA&gt; 6.530283e-24 7.310165e-20
5  ENST00000368955.7            &lt;NA&gt;     &lt;NA&gt; 6.530283e-24 7.310165e-20
6  ENST00000479258.5            &lt;NA&gt;     &lt;NA&gt; 6.842161e-24 7.310165e-20
7  ENST00000488524.1 ENSG00000115386    REG1A 1.000734e-23 9.164437e-20
8  ENST00000619379.1 ENSG00000187908    DMBT1 1.460252e-23 9.360802e-20
9  ENST00000338354.7            &lt;NA&gt;     &lt;NA&gt; 1.460252e-23 9.360802e-20
10 ENST00000368909.7            &lt;NA&gt;     &lt;NA&gt; 1.460252e-23 9.360802e-20
11 ENST00000296695.9            &lt;NA&gt;     &lt;NA&gt; 1.589414e-22 9.262528e-19
12 ENST00000321348.8            &lt;NA&gt;     &lt;NA&gt; 4.981457e-22 2.459365e-18
13 ENST00000393080.8            &lt;NA&gt;     &lt;NA&gt; 4.987480e-22 2.459365e-18
14 ENST00000427888.2 ENSG00000239754      CFB 8.326978e-22 3.139956e-18
15 ENST00000433503.2 ENSG00000241534      CFB 8.326978e-22 3.139956e-18
16 ENST00000436692.2 ENSG00000243570      CFB 8.326978e-22 3.139956e-18
17 ENST00000419411.6            &lt;NA&gt;     &lt;NA&gt; 8.326978e-22 3.139956e-18
18 ENST00000039007.4 ENSG00000036473      OTC 1.264826e-21 4.504466e-18
19 ENST00000330871.3 ENSG00000184557    SOCS3 1.683388e-21 5.395596e-18
20 ENST00000454188.5            &lt;NA&gt;     &lt;NA&gt; 1.667167e-21 5.395596e-18
   test_stat        rss degrees_free  mean_obs    var_obs     tech_var
1  113.39923  18.108697            3  7.374335  0.6873078 0.0009263481
2  112.57818 109.874120            3  9.603141  3.8569654 0.0018507640
3  112.57818 109.874120            3  9.603141  3.8569654 0.0018507640
4  111.04733 120.195423            3 10.099726  4.0904996 0.0008850097
5  111.04733 120.195423            3 10.099726  4.0904996 0.0008850097
6  110.95319 426.047794            3  3.329609 14.9232747 0.0085927194
7  110.18594 282.302693            3  2.580435  9.4763270 0.0375243358
8  109.42337 109.503692            3  9.620689  4.0154298 0.0011393785
9  109.42337 109.503692            3  9.620689  4.0154298 0.0011393785
10 109.42337 109.503692            3  9.620689  4.0154298 0.0011393785
11 104.60447  89.273925            3  6.167559  3.3852884 0.0073218897
12 102.29788  17.807119            3  9.696337  0.9522102 0.0003139664
13 102.29544 263.748235            3  6.419978 10.1830174 0.1513374771
14 101.26032  13.505785            3  7.507109  0.7457563 0.0002046838
15 101.26032  13.505785            3  7.507109  0.7457563 0.0002046838
16 101.26032  13.505785            3  7.507109  0.7457563 0.0002046838
17 101.26032  13.505785            3  7.507109  0.7457563 0.0002046838
18 100.41607   9.254886            3  8.157050  0.5746133 0.0003565220
19  99.83867 183.212473            3  5.253434  6.0925675 0.0542978318
20  99.85823 228.748372            3  2.118858  7.7303490 0.0641336692
     sigma_sq smooth_sigma_sq final_sigma_sq
1   0.7536027      0.02064112      0.7536027
2   4.5762376      0.01835476      4.5762376
3   4.5762376      0.01835476      4.5762376
4   5.0072576      0.02075054      5.0072576
5   5.0072576      0.02075054      5.0072576
6  17.7433987      0.22252250     17.7433987
7  11.7250879      0.40472948     11.7250879
8   4.5615144      0.01841836      4.5615144
9   4.5615144      0.01841836      4.5615144
10  4.5615144      0.01841836      4.5615144
11  3.7124250      0.03337283      3.7124250
12  0.7416493      0.01870914      0.7416493
13 10.8381723      0.02959264     10.8381723
14  0.5625364      0.01988880      0.5625364
15  0.5625364      0.01988880      0.5625364
16  0.5625364      0.01988880      0.5625364
17  0.5625364      0.01988880      0.5625364
18  0.3852637      0.01747636      0.3852637
19  7.5795552      0.05575429      7.5795552
20  9.4670485      0.58534292      9.4670485</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save result table</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(sleuth_table_tx, <span class="at">file =</span> <span class="st">"./outputs/01_diffexp_sleuth-transcripts-qval001-stim.csv"</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Run <code>sleuth_live(so)</code> for quick exploration of the results.</p>
<section id="visualise-transcript-counts-across-multiple-conditions" class="level4">
<h4 class="anchored" data-anchor-id="visualise-transcript-counts-across-multiple-conditions">Visualise transcript counts across multiple conditions</h4>
<p><strong>dittoSeq</strong> requires samples as colnames and targets (i.e.&nbsp;genes or transcripts) as rownames. These can be pulled from the <strong>sleuth</strong> object and wrangled to the compatible format using <strong>tidyverse</strong>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>abundances <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(so[[<span class="st">"obs_raw"</span>]])</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(abundances)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          target_id est_counts  eff_len  len       tpm  sample
1 ENST00000000233.9   2056.551 825.3027 1103  63.29273 A2-7-A1
2 ENST00000000233.9   3124.549 827.4371 1103  88.91565 A2-7-A2
3 ENST00000000233.9   3306.509 824.7783 1103 112.52427 A2-7-A3
4 ENST00000000233.9   2727.306 819.5583 1103  93.71084 A2-7-A4
5 ENST00000000233.9   2578.714 804.5703 1103  95.53383 A2-7-B1
6 ENST00000000233.9   2342.839 808.3872 1103  73.72855 A2-7-B2</code></pre>
</div>
</div>
<p>This is to create an <code>abundances</code> table with added gene annotations from <code>ttg</code> so we can see gene names in <strong>dittoSeq</strong> heatmaps.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>abundances.gene <span class="ot">&lt;-</span> <span class="fu">left_join</span>(abundances, ttg, <span class="at">by =</span> <span class="st">"target_id"</span>)</span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(abundances.gene)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          target_id est_counts  eff_len  len       tpm  sample ens_gene
1 ENST00000000233.9   2056.551 825.3027 1103  63.29273 A2-7-A1     &lt;NA&gt;
2 ENST00000000233.9   3124.549 827.4371 1103  88.91565 A2-7-A2     &lt;NA&gt;
3 ENST00000000233.9   3306.509 824.7783 1103 112.52427 A2-7-A3     &lt;NA&gt;
4 ENST00000000233.9   2727.306 819.5583 1103  93.71084 A2-7-A4     &lt;NA&gt;
5 ENST00000000233.9   2578.714 804.5703 1103  95.53383 A2-7-B1     &lt;NA&gt;
6 ENST00000000233.9   2342.839 808.3872 1103  73.72855 A2-7-B2     &lt;NA&gt;
  ext_gene
1     &lt;NA&gt;
2     &lt;NA&gt;
3     &lt;NA&gt;
4     &lt;NA&gt;
5     &lt;NA&gt;
6     &lt;NA&gt;</code></pre>
</div>
</div>
<p>Chain <code>target_id</code> (ensembl transcript ID with version) with <code>ext_gene</code> (gene symbol).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>abundances.gene<span class="sc">$</span>target_id.gene <span class="ot">&lt;-</span> <span class="fu">paste</span>(abundances.gene<span class="sc">$</span>target_id, abundances.gene<span class="sc">$</span>ext_gene, <span class="at">sep =</span> <span class="st">"_"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="co"># transcript-per-million values</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>tpm <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(abundances.gene<span class="sc">$</span>target_id.gene, abundances.gene<span class="sc">$</span>tpm, abundances.gene<span class="sc">$</span>sample) <span class="sc">%&gt;%</span> <span class="fu">setNames</span>(<span class="fu">c</span>(<span class="st">"target_id"</span>, <span class="st">"tpm"</span>, <span class="st">"sample"</span>))</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>tpm <span class="ot">&lt;-</span> tpm <span class="sc">%&gt;%</span> <span class="fu">pivot_wider</span>(<span class="at">names_from=</span>sample, <span class="at">values_from=</span>tpm) <span class="sc">%&gt;%</span> <span class="fu">data.frame</span>(<span class="at">check.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(tpm) <span class="ot">&lt;-</span> tpm<span class="sc">$</span>target_id</span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>tpm <span class="ot">&lt;-</span> <span class="fu">subset</span>(tpm, <span class="at">select=</span><span class="sc">-</span><span class="fu">c</span>(target_id))</span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(tpm[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                         A2-7-A1  A2-7-A2   A2-7-A3  A2-7-A4   A2-7-B1
ENST00000000233.9_NA    63.29273 88.91565 112.52427 93.71084  95.53383
ENST00000000412.7_NA    49.49052 45.63919  52.03826 58.57941  41.99059
ENST00000000442.10_NA   21.20029 22.52228  29.66861 18.40975  44.43467
ENST00000001008.5_FKBP4 31.63688 37.36310  40.44536 44.48072  23.40765
ENST00000001146.6_NA    47.67711 25.44619  44.64982 28.14402 132.59824</code></pre>
</div>
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="co"># log(tpm+1) values</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>logtpm <span class="ot">&lt;-</span> <span class="fu">log2</span>(tpm <span class="sc">+</span> <span class="dv">1</span>)</span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(logtpm[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                         A2-7-A1  A2-7-A2  A2-7-A3  A2-7-A4  A2-7-B1
ENST00000000233.9_NA    6.006584 6.490500 6.826857 6.565458 6.592963
ENST00000000412.7_NA    5.657941 5.543471 5.728961 5.896742 5.425949
ENST00000000442.10_NA   4.472506 4.555956 4.938691 4.278710 5.505722
ENST00000001008.5_FKBP4 5.028431 5.261647 5.373139 5.507183 4.609262
ENST00000001146.6_NA    5.605172 4.724988 5.512537 4.865128 7.061757</code></pre>
</div>
</div>
<p><strong>dittoSeq</strong> can also incorporate metadata information. These need to be a dataframe in the same order as <code>colnames</code> of the TPM dataframe.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co"># not in order of TPM dataframe</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>metadata.ditto <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(so[[<span class="st">"sample_to_covariates"</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ordered to match TPM and log(tpm+1) dataframes</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>metadata.ditto <span class="ot">&lt;-</span> metadata.ditto[<span class="fu">match</span>(<span class="fu">colnames</span>(tpm), metadata.ditto<span class="sc">$</span>sample),]</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(tpm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "A2-7-A1"    "A2-7-A2"    "A2-7-A3"    "A2-7-A4"    "A2-7-B1"   
 [6] "A2-7-B2"    "A2-7-B3"    "A2-7-B4"    "A2-7-C1"    "A2-7-C2"   
[11] "A2-7-C3"    "A2-7-C4"    "A2-7-D1"    "A2-7-D2"    "A2-7-D3"   
[16] "A2-7-D4"    "A2_11_A1"   "A2_11_A2"   "A2_11_A3"   "A2_11_A4"  
[21] "A2_11_B1"   "A2_11_B2"   "A2_11_B3"   "A2_11_B4"   "A2_12_C1"  
[26] "A2_12_C2"   "A2_12_C3"   "A2_12_C4"   "Ctrl_24"    "IL22_24"   
[31] "IL22_26_24" "IL26_24"   </code></pre>
</div>
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>metadata.ditto<span class="sc">$</span>sample</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "A2-7-A1"    "A2-7-A2"    "A2-7-A3"    "A2-7-A4"    "A2-7-B1"   
 [6] "A2-7-B2"    "A2-7-B3"    "A2-7-B4"    "A2-7-C1"    "A2-7-C2"   
[11] "A2-7-C3"    "A2-7-C4"    "A2-7-D1"    "A2-7-D2"    "A2-7-D3"   
[16] "A2-7-D4"    "A2_11_A1"   "A2_11_A2"   "A2_11_A3"   "A2_11_A4"  
[21] "A2_11_B1"   "A2_11_B2"   "A2_11_B3"   "A2_11_B4"   "A2_12_C1"  
[26] "A2_12_C2"   "A2_12_C3"   "A2_12_C4"   "Ctrl_24"    "IL22_24"   
[31] "IL22_26_24" "IL26_24"   </code></pre>
</div>
</div>
<p>Import data for <strong>dittoSeq</strong> which uses <code>SingleCellExperiment</code> structure.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>bulkSCE <span class="ot">&lt;-</span> <span class="fu">importDittoBulk</span>(<span class="at">x =</span> <span class="fu">list</span>(<span class="at">counts =</span> tpm, <span class="at">logcounts =</span> logtpm),</span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>                           <span class="at">metadata =</span> metadata.ditto)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Obtain <strong>sleuth</strong> differentially-expressed transcripts.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>tx.diff <span class="ot">&lt;-</span> <span class="fu">na.omit</span>(sleuth_table_tx)</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>tx.diff <span class="ot">&lt;-</span> <span class="fu">paste</span>(tx.diff<span class="sc">$</span>target_id, tx.diff<span class="sc">$</span>ext_gene, <span class="at">sep =</span> <span class="st">"_"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="co"># By default, dittoHeatmap prioritises "logcounts" over "counts"</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dittoHeatmap</span>(bulkSCE, tx.diff,</span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">annot.by =</span> <span class="fu">c</span>(<span class="st">"stim"</span>, <span class="st">"time"</span>, <span class="st">"medium"</span>, <span class="st">"donor"</span>), </span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a>             <span class="at">order.by =</span> (<span class="st">"stim"</span>), <span class="at">cluster_cols =</span> <span class="cn">FALSE</span>, </span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a>             <span class="at">show_rownames =</span> <span class="cn">FALSE</span>,</span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a>             <span class="at">main =</span> <span class="st">"Differential genes (qval&lt;0.01) | All conditions"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-30-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="co"># By default, dittoHeatmap prioritises "logcounts" over "counts"</span></span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dittoHeatmap</span>(bulkSCE, tx.diff,</span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">annot.by =</span> <span class="fu">c</span>(<span class="st">"stim"</span>, <span class="st">"time"</span>, <span class="st">"batch"</span>), </span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a>             <span class="at">order.by =</span> (<span class="st">"stim"</span>), <span class="at">cluster_cols =</span> <span class="cn">FALSE</span>, </span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a>             <span class="at">show_rownames =</span> <span class="cn">FALSE</span>,</span>
<span id="cb65-6"><a href="#cb65-6" aria-hidden="true" tabindex="-1"></a>             <span class="at">main =</span> <span class="st">"Differential genes (qval&lt;0.01) | All conditions"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-31-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>IL22 or IL22+IL26 stimulation has a distinct effect with ~1000 differentially-expressed genes (qval &lt;= 0.01). At this resolution IL26 and control (unstimulated) look fairly similar.</p>
</section>
<section id="go-analysis---panther-classification-system" class="level4">
<h4 class="anchored" data-anchor-id="go-analysis---panther-classification-system">GO analysis - PANTHER classification system</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rba_panther_info</span>(<span class="at">what=</span><span class="st">"datasets"</span>) <span class="co"># list dataset IDs</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  release_date
1   2025-07-22
2   2025-07-22
3   2025-07-22
4   2024-06-20
5   2024-06-20
6   2024-06-20
7   2024-06-20
8   2024-06-20
9   2023-09-07
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          description
1                                                                                                                                                                                                                                                                                                                                                                                                                            Gene Ontology Molecular Function annotations including both manually curated and electronic annotations.
2                                                                                                                                                                                                                                                                                                                                                                                                                            Gene Ontology Biological Process annotations including both manually curated and electronic annotations.
3                                                                                                                                                                                                                                                                                                                                                                                                                            Gene Ontology Cellular Component annotations including both manually curated and electronic annotations.
4                                                                                                                                                                                                                                                           A molecular process that can be carried out by the action of a single macromolecular machine, usually via direct physical interactions with other molecular entities. Function in this sense denotes an action, or activity, that a gene product (or a complex) performs.
5                                                                                                       A biological process is the execution of a genetically-encoded biological module or program. It consists of all the steps required to achieve the specific biological objective of the module. A biological process is accomplished by a particular set of molecular functions carried out by specific gene products (or macromolecular complexes), often in a highly regulated manner and in a particular temporal sequence.
6 A location, relative to cellular compartments and structures, occupied by a macromolecular machine. There are three types of cellular components described in the gene ontology: (1) the cellular anatomical entity where a gene product carries out a molecular function (e.g., plasma membrane, cytoskeleton) or membrane-enclosed compartments (e.g., mitochondrion); (2) virion components, where viral proteins act, and (3) the stable macromolecular complexes of which gene product are parts (e.g., the clathrin complex).
7                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    
8                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    Panther Pathways
9                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   Reactome Pathways
                                id                              label
1                       GO:0003674                 molecular_function
2                       GO:0008150                 biological_process
3                       GO:0005575                 cellular_component
4 ANNOT_TYPE_ID_PANTHER_GO_SLIM_MF PANTHER GO Slim Molecular Function
5 ANNOT_TYPE_ID_PANTHER_GO_SLIM_BP PANTHER GO Slim Biological Process
6 ANNOT_TYPE_ID_PANTHER_GO_SLIM_CC  PANTHER GO Slim Cellular Location
7         ANNOT_TYPE_ID_PANTHER_PC                      protein class
8    ANNOT_TYPE_ID_PANTHER_PATHWAY         ANNOT_TYPE_PANTHER_PATHWAY
9   ANNOT_TYPE_ID_REACTOME_PATHWAY        ANNOT_TYPE_REACTOME_PATHWAY
                  version
1 10.5281/zenodo.16423886
2 10.5281/zenodo.16423886
3 10.5281/zenodo.16423886
4                      19
5                      19
6                      19
7                      19
8                      19
9                      86</code></pre>
</div>
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">rba_panther_info</span>(<span class="at">what=</span><span class="st">"organisms"</span>))  <span class="co"># list organism IDs</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       name taxon_id short_name                    version
1     human     9606      HUMAN Reference Proteome 2023_03
2     mouse    10090      MOUSE Reference Proteome 2023_03
3       rat    10116        RAT Reference Proteome 2023_03
4   chicken     9031      CHICK Reference Proteome 2023_03
5 zebrafish     7955      DANRE Reference Proteome 2023_03
6 fruit_fly     7227      DROME Reference Proteome 2023_03
                long_name
1            Homo sapiens
2            Mus musculus
3       Rattus norvegicus
4           Gallus gallus
5             Danio rerio
6 Drosophila melanogaster</code></pre>
</div>
</div>
<p>Get gene list-of-interest i.e.&nbsp;differentially-expressed genes (up- or down-regulated) in response to IL stimulation.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>genes.diff <span class="ot">&lt;-</span> sleuth_table_gene<span class="sc">$</span>ext_gene</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Perform Fisher’s exact test with FDR &lt; 0.01, Reactome Pathways annotation.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>go <span class="ot">&lt;-</span> <span class="fu">rba_panther_enrich</span>(<span class="at">genes =</span> genes.diff,</span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>                          <span class="at">organism =</span> <span class="dv">9606</span> , <span class="at">annot_dataset =</span> <span class="st">"ANNOT_TYPE_ID_REACTOME_PATHWAY"</span>,</span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a>                          <span class="at">cutoff =</span> <span class="fl">0.01</span>,</span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a>                          <span class="at">test_type =</span> <span class="st">"FISHER"</span>,</span>
<span id="cb71-5"><a href="#cb71-5" aria-hidden="true" tabindex="-1"></a>                          <span class="at">correction =</span> <span class="st">"FDR"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Performing PANTHER over-representation analysis (Fisher's exact test) on 1343 genes from `organism 9606` against `ANNOT_TYPE_ID_REACTOME_PATHWAY` datasets.</code></pre>
</div>
</div>
<p>Plot as diverging barchart.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>go.res <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(go<span class="sc">$</span>result)</span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>go.res<span class="sc">$</span>GO <span class="ot">&lt;-</span> <span class="fu">paste</span>(go.res<span class="sc">$</span>term.id, go.res<span class="sc">$</span>term.label,<span class="at">sep=</span><span class="st">"_"</span>)</span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>go.res<span class="sc">$</span>representation <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(go.res<span class="sc">$</span>fold_enrichment <span class="sc">&lt;</span> <span class="dv">1</span>, <span class="st">"under"</span>, <span class="st">"over"</span>)</span>
<span id="cb73-4"><a href="#cb73-4" aria-hidden="true" tabindex="-1"></a>go.res <span class="ot">&lt;-</span> go.res[<span class="fu">order</span>(<span class="sc">-</span>go.res<span class="sc">$</span>fold_enrichment), ]</span>
<span id="cb73-5"><a href="#cb73-5" aria-hidden="true" tabindex="-1"></a><span class="co"># convert to factor to retain sorted order in plot</span></span>
<span id="cb73-6"><a href="#cb73-6" aria-hidden="true" tabindex="-1"></a>go.res<span class="sc">$</span>GO <span class="ot">&lt;-</span> <span class="fu">factor</span>(go.res<span class="sc">$</span>GO, <span class="at">levels =</span> go.res<span class="sc">$</span>GO)</span>
<span id="cb73-7"><a href="#cb73-7" aria-hidden="true" tabindex="-1"></a>go.res <span class="ot">&lt;-</span> go.res <span class="sc">%&gt;%</span> <span class="fu">filter</span>(fold_enrichment <span class="sc">&gt;</span> <span class="dv">1</span>)</span>
<span id="cb73-8"><a href="#cb73-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-9"><a href="#cb73-9" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(go.res, <span class="fu">aes</span>(<span class="at">x=</span>GO, <span class="at">y=</span>fold_enrichment, <span class="at">label=</span>fold_enrichment)) <span class="sc">+</span> </span>
<span id="cb73-10"><a href="#cb73-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">'identity'</span>, <span class="fu">aes</span>(<span class="at">fill=</span>representation), <span class="at">width=</span>.<span class="dv">5</span>)  <span class="sc">+</span></span>
<span id="cb73-11"><a href="#cb73-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">name=</span><span class="st">"Representation"</span>, </span>
<span id="cb73-12"><a href="#cb73-12" aria-hidden="true" tabindex="-1"></a>                    <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"Over"</span>, <span class="st">"Under"</span>), </span>
<span id="cb73-13"><a href="#cb73-13" aria-hidden="true" tabindex="-1"></a>                    <span class="at">values =</span> <span class="fu">c</span>(<span class="st">"over"</span><span class="ot">=</span><span class="st">"#0072B2"</span>, <span class="st">"under"</span><span class="ot">=</span><span class="st">"#D55E00"</span>)) <span class="sc">+</span> </span>
<span id="cb73-14"><a href="#cb73-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">subtitle=</span><span class="st">"PANTHER statistical overrepresentation test | Fisher's exact | FDR &lt; 0.01 | fold_enrichment &gt; 1"</span>, </span>
<span id="cb73-15"><a href="#cb73-15" aria-hidden="true" tabindex="-1"></a>       <span class="at">title=</span> <span class="st">"Differential genes qval &lt; 0.01"</span>) <span class="sc">+</span> </span>
<span id="cb73-16"><a href="#cb73-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.subtitle=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">5</span>, <span class="at">face=</span><span class="st">'italic'</span>)) <span class="sc">+</span></span>
<span id="cb73-17"><a href="#cb73-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">breaks =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">13</span>, <span class="at">by =</span> <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb73-18"><a href="#cb73-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_flip</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-35-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Does IL26 have a(ny) subtle effect masked by this analysis? Pairwise test to explore.</p>
</section>
</section>
<section id="test-for-differentially-expressed-genes---pairwise-il26-vs-control" class="level3">
<h3 class="anchored" data-anchor-id="test-for-differentially-expressed-genes---pairwise-il26-vs-control">Test for differentially-expressed genes - pairwise (IL26 vs control)</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>metadata.il26 <span class="ot">&lt;-</span> <span class="fu">subset</span>(metadata, stim <span class="sc">==</span> <span class="st">"IL26"</span> <span class="sc">|</span> stim <span class="sc">==</span> <span class="st">"ctl"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>so.il26 <span class="ot">&lt;-</span> <span class="fu">sleuth_prep</span>(metadata.il26, <span class="at">target_mapping =</span> ttg, <span class="at">aggregation_column =</span> <span class="st">'ens_gene'</span>, </span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>                       <span class="at">extra_bootstrap_summary =</span> <span class="cn">TRUE</span>, </span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a>                       <span class="at">read_bootstrap_tpm =</span> <span class="cn">TRUE</span>) <span class="co"># turn this flag on to plot TPM bootstaps later on</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save sleuth object to reload without recomputing bootstraps</span></span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(so.il26, <span class="at">file =</span> <span class="st">"./01_diffexp_sleuth-so.il26.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>so.il26 <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"./01_diffexp_sleuth-so.il26.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="co"># covariates</span></span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(so.il26, <span class="at">units =</span> <span class="st">"tpm"</span>, <span class="at">color_by =</span> <span class="st">"donor"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-40-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(so.il26, <span class="at">units =</span> <span class="st">"tpm"</span>, <span class="at">color_by =</span> <span class="st">"batch"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-40-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(so.il26, <span class="at">units =</span> <span class="st">"tpm"</span>, <span class="at">color_by =</span> <span class="st">"medium"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-40-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="co"># conditions</span></span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(so.il26, <span class="at">units =</span> <span class="st">"tpm"</span>, <span class="at">color_by =</span> <span class="st">"time"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-41-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(so.il26, <span class="at">units =</span> <span class="st">"tpm"</span>, <span class="at">color_by =</span> <span class="st">"stim"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-41-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Fit a <code>reduced</code> model, which includes the parameters corresponding to experimental covariates we are controlling for i.e.&nbsp;<em>batch</em> (which will account for <em>medium</em>, <em>time</em>) and <em>donor</em>. At this point <em>time</em> is accounted for i.e.&nbsp;to control for genes that may be differentially expressed at different timepoints independently of <em>stim</em>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a>so.il26 <span class="ot">&lt;-</span> <span class="fu">sleuth_fit</span>(so.il26, <span class="sc">~</span>batch <span class="sc">+</span> donor, <span class="st">'reduced'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>fitting measurement error models</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>shrinkage estimation</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>3 NA values were found during variance shrinkage estimation due to mean observation values outside of the range used for the LOESS fit.
The LOESS fit will be repeated using exact computation of the fitted surface to extrapolate the missing values.
These are the target ids with NA values: ENST00000512751.5, ENST00000356495.4, ENST00000361624.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>computing variance of betas</code></pre>
</div>
</div>
<p>Fit the <code>full</code> model which includes all parameters including <em>stim</em>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a>so.il26 <span class="ot">&lt;-</span> <span class="fu">sleuth_fit</span>(so.il26, <span class="sc">~</span>batch <span class="sc">+</span> donor <span class="sc">+</span> stim, <span class="st">'full'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>fitting measurement error models</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>shrinkage estimation</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>3 NA values were found during variance shrinkage estimation due to mean observation values outside of the range used for the LOESS fit.
The LOESS fit will be repeated using exact computation of the fitted surface to extrapolate the missing values.
These are the target ids with NA values: ENST00000512751.5, ENST00000356495.4, ENST00000361624.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>computing variance of betas</code></pre>
</div>
</div>
<p>Then compare the <code>full</code> model to the <code>reduced</code> model with the likelihood ratio test. Using the likelihood ratio test, sleuth identifies genes whose abundances are significantly better explained when IL26 treatment (<em>stim</em>) is taken into account, while accounting for baseline differences that may be explained by <em>batch</em> or <em>donor</em>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a>so.il26 <span class="ot">&lt;-</span> <span class="fu">sleuth_lrt</span>(so.il26, <span class="st">'reduced'</span>, <span class="st">'full'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Obtain gene-level differential expression results. Sleuth uses the p-values from comparing transcripts to make a gene-level determination and perform gene differential expression.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a>sleuth_table_gene.il26 <span class="ot">&lt;-</span> <span class="fu">sleuth_results</span>(so.il26, <span class="st">'reduced:full'</span>, <span class="st">'lrt'</span>, <span class="at">show_all =</span> <span class="cn">FALSE</span>)</span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a>sleuth_table_gene.il26 <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(sleuth_table_gene.il26, qval <span class="sc">&lt;=</span> <span class="fl">0.1</span>) <span class="co"># zero genes &lt;= 0.05 </span></span>
<span id="cb94-3"><a href="#cb94-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sleuth_table_gene.il26, <span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        target_id ext_gene num_aggregated_transcripts sum_mean_obs_counts
1 ENSG00000187908    DMBT1                          1            7.869715
2 ENSG00000183049   CAMK1D                          1            1.479010
3 ENSG00000187193     MT1X                          2            7.918024
          pval        qval
1 7.104320e-07 0.008928709
2 5.970526e-06 0.037518783
3 1.800217e-05 0.075417094</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save result table</span></span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(sleuth_table_gene.il26, <span class="at">file =</span> <span class="st">"./outputs/01_diffexp_sleuth-genes-qval01-IL26.csv"</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Obtain transcript-level differential expression results.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a>sleuth_table_tx.il26 <span class="ot">&lt;-</span> <span class="fu">sleuth_results</span>(so.il26, <span class="st">'reduced:full'</span>, <span class="st">'lrt'</span>, <span class="at">show_all =</span> <span class="cn">FALSE</span>, <span class="at">pval_aggregate =</span> <span class="cn">FALSE</span>)</span>
<span id="cb97-2"><a href="#cb97-2" aria-hidden="true" tabindex="-1"></a>sleuth_table_tx.il26 <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(sleuth_table_tx.il26, qval <span class="sc">&lt;=</span> <span class="fl">0.1</span>)</span>
<span id="cb97-3"><a href="#cb97-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sleuth_table_tx.il26, <span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            target_id        ens_gene ext_gene         pval         qval
1   ENST00000330163.8            &lt;NA&gt;     &lt;NA&gt; 1.079840e-09 3.545222e-05
2   ENST00000368956.6            &lt;NA&gt;     &lt;NA&gt; 1.079840e-09 3.545222e-05
3   ENST00000619379.1 ENSG00000187908    DMBT1 7.104320e-07 9.329677e-03
4   ENST00000338354.7            &lt;NA&gt;     &lt;NA&gt; 7.104320e-07 9.329677e-03
5   ENST00000368909.7            &lt;NA&gt;     &lt;NA&gt; 7.104320e-07 9.329677e-03
6   ENST00000344338.7            &lt;NA&gt;     &lt;NA&gt; 1.523934e-06 1.429494e-02
7   ENST00000368955.7            &lt;NA&gt;     &lt;NA&gt; 1.523934e-06 1.429494e-02
8   ENST00000464744.5            &lt;NA&gt;     &lt;NA&gt; 2.723463e-06 2.235350e-02
9   ENST00000394634.6            &lt;NA&gt;     &lt;NA&gt; 3.303491e-06 2.410154e-02
10 ENST00000311713.11            &lt;NA&gt;     &lt;NA&gt; 4.425090e-06 2.641457e-02
11  ENST00000449081.6            &lt;NA&gt;     &lt;NA&gt; 4.196372e-06 2.641457e-02
12  ENST00000394485.4 ENSG00000187193     MT1X 5.488730e-06 3.003342e-02
13  ENST00000615792.1 ENSG00000183049   CAMK1D 5.970526e-06 3.015667e-02
14 ENST00000320443.12            &lt;NA&gt;     &lt;NA&gt; 8.036347e-06 3.769161e-02
15  ENST00000378254.5            &lt;NA&gt;     &lt;NA&gt; 9.730455e-06 3.993257e-02
16  ENST00000541428.4            &lt;NA&gt;     &lt;NA&gt; 9.286272e-06 3.993257e-02
17 ENST00000221992.10            &lt;NA&gt;     &lt;NA&gt; 1.385628e-05 5.054617e-02
18  ENST00000421699.6            &lt;NA&gt;     &lt;NA&gt; 1.368563e-05 5.054617e-02
19  ENST00000381090.7            &lt;NA&gt;     &lt;NA&gt; 1.779247e-05 6.103538e-02
20  ENST00000465241.5            &lt;NA&gt;     &lt;NA&gt; 1.859078e-05 6.103538e-02
   test_stat        rss degrees_free mean_obs   var_obs     tech_var   sigma_sq
1   37.17510  5.0550148            1 7.840726 0.8370407 0.0035913776 0.62828547
2   37.17510  5.0550148            1 7.840726 0.8370407 0.0035913776 0.62828547
3   24.58662  6.5212262            1 7.869715 1.2514294 0.0022301439 0.81292314
4   24.58662  6.5212262            1 7.869715 1.2514294 0.0022301439 0.81292314
5   24.58662  6.5212262            1 7.869715 1.2514294 0.0022301439 0.81292314
6   23.11755  6.7367582            1 8.258614 0.8392234 0.0017360384 0.84035873
7   23.11755  6.7367582            1 8.258614 0.8392234 0.0017360384 0.84035873
8   22.00214 56.1240536            1 1.888006 4.4992793 1.2878328125 5.72767388
9   21.63169 15.7341039            1 4.887342 3.2478790 0.2655098686 1.70125312
10  21.07129 37.1823221            1 2.011316 3.7251295 0.6462984210 4.00149184
11  21.17299  2.2002072            1 6.233650 0.3377679 0.0594875139 0.21553839
12  20.65870  0.8292244            1 6.040390 0.9146518 0.0035452507 0.10010780
13  20.49763 55.3188111            1 1.479010 4.0430102 1.2376164895 5.67723490
14  19.92920 55.2587300            1 2.319572 4.6075060 2.2527884834 4.65455276
15  19.56362 41.7547558            1 1.587982 3.5389113 1.1032541816 4.11609029
16  19.65289  0.1989220            1 7.996155 0.1073590 0.0011051674 0.02376009
17  18.88881  0.2998424            1 9.545262 1.5174246 0.0004137942 0.03706650
18  18.91245  1.3934287            1 7.124160 0.2194301 0.0224273080 0.15175128
19  18.41210 51.8498522            1 1.564771 4.6270047 1.4257861653 5.05544536
20  18.32847 33.4638934            1 1.373187 3.3045897 0.6376328102 3.54535387
   smooth_sigma_sq final_sigma_sq
1      0.010804107     0.62828547
2      0.010804107     0.62828547
3      0.010711065     0.81292314
4      0.010711065     0.81292314
5      0.010711065     0.81292314
6      0.009709935     0.84035873
7      0.009709935     0.84035873
8      0.682055213     5.72767388
9      0.049921531     1.70125312
10     0.614152529     4.00149184
11     0.021972826     0.21553839
12     0.024531391     0.10010780
13     0.956917444     5.67723490
14     0.470594924     4.65455276
15     0.875551762     4.11609029
16     0.010336469     0.02376009
17     0.009078699     0.03706650
18     0.014115561     0.15175128
19     0.892351381     5.05544536
20     1.042266636     3.54535387</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save result table</span></span>
<span id="cb99-2"><a href="#cb99-2" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(sleuth_table_tx.il26, <span class="at">file =</span> <span class="st">"./outputs/01_diffexp_sleuth-transcripts-qval01-IL26.csv"</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Three genes differentially-expressed (qval &lt; 0.1) in response to IL26 stimulation. Visualise how each is expressed across the different conditions and covariates by plotting bootstraps per transcript.</p>
<ul>
<li><strong>CAMK1D</strong> slightly upregulated (qval = 0.038) in IL26 samples…</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_bootstrap</span>(so.il26, <span class="st">"ENST00000615792.1"</span>, <span class="at">units =</span> <span class="st">"tpm"</span>, <span class="at">color_by =</span> <span class="st">"stim"</span>) <span class="co"># CAMK1D transcript</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-49-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>…This observation is consistent across donors…</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_bootstrap</span>(so.il26, <span class="st">"ENST00000615792.1"</span>, <span class="at">units =</span> <span class="st">"tpm"</span>, <span class="at">color_by =</span> <span class="st">"donor"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-50-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>…and batches, except for batch 2 which comes from donor TIP-OX-135…</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_bootstrap</span>(so.il26, <span class="st">"ENST00000615792.1"</span>, <span class="at">units =</span> <span class="st">"tpm"</span>, <span class="at">color_by =</span> <span class="st">"batch"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-51-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>…and appears slightly elevated in 48h samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_bootstrap</span>(so.il26, <span class="st">"ENST00000615792.1"</span>, <span class="at">units =</span> <span class="st">"tpm"</span>, <span class="at">color_by =</span> <span class="st">"time"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-52-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<ul>
<li><strong>DMBT1</strong> upregulated (qval = 0.008) in <em>some</em> IL26 samples…</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_bootstrap</span>(so.il26, <span class="st">"ENST00000619379.1"</span>, <span class="at">units =</span> <span class="st">"tpm"</span>, <span class="at">color_by =</span> <span class="st">"stim"</span>) <span class="co"># DMBT1 transcript</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-53-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>…observed in two different donors…</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_bootstrap</span>(so.il26, <span class="st">"ENST00000619379.1"</span>, <span class="at">units =</span> <span class="st">"tpm"</span>, <span class="at">color_by =</span> <span class="st">"donor"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-54-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>…from batch four i.e.&nbsp;48h timepoint.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_bootstrap</span>(so.il26, <span class="st">"ENST00000619379.1"</span>, <span class="at">units =</span> <span class="st">"tpm"</span>, <span class="at">color_by =</span> <span class="st">"time"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-55-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_bootstrap</span>(so.il26, <span class="st">"ENST00000619379.1"</span>, <span class="at">units =</span> <span class="st">"tpm"</span>, <span class="at">color_by =</span> <span class="st">"batch"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-56-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<ul>
<li>Put into context of the earlier multi-condition analysis.</li>
</ul>
<div class="cell">
<div class="sourceCode cell-code" id="cb108"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a><span class="co"># By default, dittoHeatmap prioritises "logcounts" over "counts"</span></span>
<span id="cb108-2"><a href="#cb108-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dittoHeatmap</span>(bulkSCE, tx.diff,</span>
<span id="cb108-3"><a href="#cb108-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">annot.by =</span> <span class="fu">c</span>(<span class="st">"stim"</span>, <span class="st">"time"</span>, <span class="st">"medium"</span>, <span class="st">"donor"</span>), </span>
<span id="cb108-4"><a href="#cb108-4" aria-hidden="true" tabindex="-1"></a>             <span class="at">order.by =</span> (<span class="st">"stim"</span>), <span class="at">cluster_cols =</span> <span class="cn">FALSE</span>, </span>
<span id="cb108-5"><a href="#cb108-5" aria-hidden="true" tabindex="-1"></a>             <span class="at">show_rownames =</span> <span class="cn">FALSE</span>,</span>
<span id="cb108-6"><a href="#cb108-6" aria-hidden="true" tabindex="-1"></a>             <span class="at">highlight.features =</span>  <span class="fu">c</span>(<span class="st">"ENST00000615792.1_CAMK1D"</span>, <span class="st">"ENST00000619379.1_DMBT1"</span>, <span class="st">"ENST00000394485.4_MT1X"</span>),</span>
<span id="cb108-7"><a href="#cb108-7" aria-hidden="true" tabindex="-1"></a>             <span class="at">main =</span> <span class="st">"Differential genes (qval&lt;0.01) | IL22/22+26/26 stim vs control"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-57-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p><strong>H Chen</strong> - IL26 might not do too much on its own but could affect IL22-mediated effects.</p>
</section>
<section id="test-for-differentially-expressed-genes---pairwise-il22il26-vs-il22" class="level3">
<h3 class="anchored" data-anchor-id="test-for-differentially-expressed-genes---pairwise-il22il26-vs-il22">Test for differentially-expressed genes - pairwise (IL22+IL26 vs IL22)</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a>metadata.il22.il26 <span class="ot">&lt;-</span> <span class="fu">subset</span>(metadata, stim <span class="sc">==</span> <span class="st">"IL22"</span> <span class="sc">|</span> stim <span class="sc">==</span> <span class="st">"IL22+IL26"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a>so.il22.il26 <span class="ot">&lt;-</span> <span class="fu">sleuth_prep</span>(metadata.il22.il26, </span>
<span id="cb110-2"><a href="#cb110-2" aria-hidden="true" tabindex="-1"></a>                            <span class="at">target_mapping =</span> ttg, </span>
<span id="cb110-3"><a href="#cb110-3" aria-hidden="true" tabindex="-1"></a>                            <span class="at">aggregation_column =</span> <span class="st">'ens_gene'</span>,</span>
<span id="cb110-4"><a href="#cb110-4" aria-hidden="true" tabindex="-1"></a>                            <span class="at">extra_bootstrap_summary =</span> <span class="cn">TRUE</span>, </span>
<span id="cb110-5"><a href="#cb110-5" aria-hidden="true" tabindex="-1"></a>                            <span class="at">read_bootstrap_tpm =</span> <span class="cn">TRUE</span>) <span class="co"># turn this flag on to plot TPM bootstraps later on</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save sleuth object to reload without recomputing bootstraps</span></span>
<span id="cb111-2"><a href="#cb111-2" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(so.il22.il26, <span class="at">file =</span> <span class="st">"./01_diffexp_sleuth-so.il22.il26.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a>so.il22.il26 <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"./01_diffexp_sleuth-so.il22.il26.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a><span class="co"># covariates</span></span>
<span id="cb113-2"><a href="#cb113-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(so.il22.il26, <span class="at">units =</span> <span class="st">"tpm"</span>, <span class="at">color_by =</span> <span class="st">"donor"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-62-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(so.il22.il26, <span class="at">units =</span> <span class="st">"tpm"</span>, <span class="at">color_by =</span> <span class="st">"batch"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-62-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb115"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(so.il22.il26, <span class="at">units =</span> <span class="st">"tpm"</span>, <span class="at">color_by =</span> <span class="st">"medium"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-62-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a><span class="co"># conditions</span></span>
<span id="cb116-2"><a href="#cb116-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(so.il22.il26, <span class="at">units =</span> <span class="st">"tpm"</span>, <span class="at">color_by =</span> <span class="st">"time"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-63-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb117"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb117-1"><a href="#cb117-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pca</span>(so.il22.il26, <span class="at">units =</span> <span class="st">"tpm"</span>, <span class="at">color_by =</span> <span class="st">"stim"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-63-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Fit a <code>reduced</code> model, which includes the parameters corresponding to experimental covariates we are controlling for i.e.&nbsp;<em>batch</em> (which will account for <em>medium</em>, <em>time</em>) and <em>donor</em>. At this point <em>time</em> is accounted for i.e.&nbsp;to control for genes that may be differentially expressed at different timepoints independently of <em>stim</em>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb118"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a>so.il22.il26 <span class="ot">&lt;-</span> <span class="fu">sleuth_fit</span>(so.il22.il26, <span class="sc">~</span>batch <span class="sc">+</span> donor, <span class="st">'reduced'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>fitting measurement error models</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>shrinkage estimation</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>2 NA values were found during variance shrinkage estimation due to mean observation values outside of the range used for the LOESS fit.
The LOESS fit will be repeated using exact computation of the fitted surface to extrapolate the missing values.
These are the target ids with NA values: ENST00000219022.2, ENST00000361624.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>computing variance of betas</code></pre>
</div>
</div>
<p>Fit the <code>full</code> model which includes all parameters including <em>stim</em>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb123"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb123-1"><a href="#cb123-1" aria-hidden="true" tabindex="-1"></a>so.il22.il26 <span class="ot">&lt;-</span> <span class="fu">sleuth_fit</span>(so.il22.il26, <span class="sc">~</span>batch <span class="sc">+</span> donor <span class="sc">+</span> stim, <span class="st">'full'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>fitting measurement error models</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>shrinkage estimation</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>2 NA values were found during variance shrinkage estimation due to mean observation values outside of the range used for the LOESS fit.
The LOESS fit will be repeated using exact computation of the fitted surface to extrapolate the missing values.
These are the target ids with NA values: ENST00000219022.2, ENST00000361624.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>computing variance of betas</code></pre>
</div>
</div>
<p>Then compare the <code>full</code> model to the <code>reduced</code> model with the likelihood ratio test. Using the likelihood ratio test, sleuth identifies genes whose abundances are significantly better explained when IL26 treatment (<em>stim</em>) is taken into account, while accounting for baseline differences that may be explained by <em>batch</em> or <em>donor</em>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb128"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb128-1"><a href="#cb128-1" aria-hidden="true" tabindex="-1"></a>so.il22.il26 <span class="ot">&lt;-</span> <span class="fu">sleuth_lrt</span>(so.il22.il26, <span class="st">'reduced'</span>, <span class="st">'full'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Obtain gene-level differential expression results. Sleuth uses the p-values from comparing transcripts to make a gene-level determination and perform gene differential expression.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb129"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb129-1"><a href="#cb129-1" aria-hidden="true" tabindex="-1"></a>sleuth_table_gene.il22.il26 <span class="ot">&lt;-</span> <span class="fu">sleuth_results</span>(so.il22.il26, <span class="st">'reduced:full'</span>, <span class="st">'lrt'</span>, <span class="at">show_all =</span> <span class="cn">FALSE</span>)</span>
<span id="cb129-2"><a href="#cb129-2" aria-hidden="true" tabindex="-1"></a>sleuth_table_gene.il22.il26 <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(sleuth_table_gene.il22.il26, qval <span class="sc">&lt;=</span> <span class="fl">0.05</span>)</span>
<span id="cb129-3"><a href="#cb129-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sleuth_table_gene.il22.il26, <span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         target_id  ext_gene num_aggregated_transcripts sum_mean_obs_counts
1  ENSG00000124570  SERPINB6                          1            5.589838
2  ENSG00000167065    DUSP18                          4           18.344722
3  ENSG00000102144      PGK1                          4           28.067188
4  ENSG00000170537      TMC7                          1            6.289153
5  ENSG00000181577  C6orf223                          3           15.464653
6  ENSG00000137673      MMP7                          2           10.976688
7  ENSG00000137440    FGFBP1                          1            7.740898
8  ENSG00000176907    C8orf4                          1            6.621298
9  ENSG00000120800     UTP20                          1            7.380428
10 ENSG00000232388 LINC00493                          2           11.436692
11 ENSG00000130066      SAT1                          3           18.734066
12 ENSG00000013588    GPRC5A                          4           28.980578
13 ENSG00000034510    TMSB10                          1           10.513427
14 ENSG00000160180      TFF3                          2           16.119184
           pval       qval
1  1.506742e-05 0.02880684
2  1.336450e-05 0.02880684
3  8.518345e-06 0.02880684
4  1.859256e-05 0.02880684
5  1.723114e-05 0.02880684
6  1.540802e-05 0.02880684
7  1.033695e-05 0.02880684
8  3.735930e-06 0.02880684
9  2.544494e-05 0.03504334
10 3.720178e-05 0.04019571
11 3.953919e-05 0.04019571
12 3.736394e-05 0.04019571
13 4.540055e-05 0.04019571
14 4.474766e-05 0.04019571</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb131"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb131-1"><a href="#cb131-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save result table</span></span>
<span id="cb131-2"><a href="#cb131-2" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(sleuth_table_gene.il22.il26, <span class="at">file =</span> <span class="st">"./outputs/01_diffexp_sleuth-genes-qval005-IL22vsIL22+26.csv"</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Obtain transcript-level differential expression results.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb132"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb132-1"><a href="#cb132-1" aria-hidden="true" tabindex="-1"></a>sleuth_table_tx.il22.il26 <span class="ot">&lt;-</span> <span class="fu">sleuth_results</span>(so.il22.il26, <span class="st">'reduced:full'</span>, <span class="st">'lrt'</span>, <span class="at">show_all =</span> <span class="cn">FALSE</span>, <span class="at">pval_aggregate =</span> <span class="cn">FALSE</span>)</span>
<span id="cb132-2"><a href="#cb132-2" aria-hidden="true" tabindex="-1"></a>sleuth_table_tx.il22.il26 <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(sleuth_table_tx.il22.il26, qval <span class="sc">&lt;=</span> <span class="fl">0.05</span>)</span>
<span id="cb132-3"><a href="#cb132-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sleuth_table_tx.il22.il26, <span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            target_id        ens_gene ext_gene         pval       qval
1   ENST00000380539.5            &lt;NA&gt;     &lt;NA&gt; 2.172516e-07 0.01404727
2   ENST00000312419.7            &lt;NA&gt;     &lt;NA&gt; 1.079265e-06 0.01853754
3  ENST00000326427.10            &lt;NA&gt;     &lt;NA&gt; 8.161081e-07 0.01853754
4   ENST00000358789.7            &lt;NA&gt;     &lt;NA&gt; 1.471270e-06 0.01853754
5   ENST00000493975.5            &lt;NA&gt;     &lt;NA&gt; 2.293576e-06 0.01853754
6   ENST00000544604.6            &lt;NA&gt;     &lt;NA&gt; 1.850652e-06 0.01853754
7   ENST00000632792.1            &lt;NA&gt;     &lt;NA&gt; 1.377323e-06 0.01853754
8   ENST00000635083.1            &lt;NA&gt;     &lt;NA&gt; 2.185486e-06 0.01853754
9   ENST00000622186.4            &lt;NA&gt;     &lt;NA&gt; 3.134801e-06 0.02252145
10  ENST00000315792.4 ENSG00000176907   C8orf4 3.735930e-06 0.02415615
11  ENST00000453385.1 ENSG00000122140    MRPS2 6.979534e-06 0.03091028
12  ENST00000292401.8            &lt;NA&gt;     &lt;NA&gt; 6.521691e-06 0.03091028
13  ENST00000297875.6            &lt;NA&gt;     &lt;NA&gt; 6.356775e-06 0.03091028
14  ENST00000377625.6            &lt;NA&gt;     &lt;NA&gt; 7.648810e-06 0.03091028
15  ENST00000379442.7            &lt;NA&gt;     &lt;NA&gt; 7.284927e-06 0.03091028
16  ENST00000443024.6            &lt;NA&gt;     &lt;NA&gt; 7.151986e-06 0.03091028
17  ENST00000508167.5            &lt;NA&gt;     &lt;NA&gt; 8.598400e-06 0.03270376
18  ENST00000372684.7            &lt;NA&gt;     &lt;NA&gt; 9.217946e-06 0.03311240
19  ENST00000382333.1 ENSG00000137440   FGFBP1 1.033695e-05 0.03409581
20  ENST00000450894.7            &lt;NA&gt;     &lt;NA&gt; 1.054635e-05 0.03409581
   test_stat        rss degrees_free mean_obs    var_obs     tech_var
1   26.87319  1.3510293            1 7.499785 0.67167626 0.0139089796
2   23.78129  0.2458086            1 7.390601 0.25722352 0.0009921625
3   24.31945  0.1668187            1 9.237607 0.24431953 0.0002921633
4   23.18518  0.1747008            1 8.997427 0.39436003 0.0001517003
5   22.33196 59.9301376            1 1.906108 4.49586474 1.5074088233
6   22.74417  0.2504223            1 7.287973 0.10099622 0.0008254028
7   23.31208 49.7962494            1 2.415906 5.17978580 1.2543755922
8   22.42467 28.0234605            1 1.448439 2.18137895 0.2742221027
9   21.73223  0.1674669            1 8.290881 0.07425807 0.0005082416
10  21.39578  0.3951005            1 6.621298 0.49693930 0.0016025725
11  20.19884 54.4286298            1 1.341978 4.92460317 1.1432127366
12  20.32864  0.1972215            1 7.207902 0.29937263 0.0008395418
13  20.37765  0.3589276            1 6.507534 0.60745362 0.0023775102
14  20.02370  0.1543288            1 7.861537 0.10268112 0.0004827137
15  20.11692 65.2910980            1 3.381789 6.63321758 1.4459131566
16  20.15215 73.9782112            1 4.062362 7.02038671 2.1752798754
17  19.79997  0.2152819            1 8.268921 0.10114301 0.0018595511
18  19.66700  0.1943582            1 7.173530 0.63132055 0.0009494448
19  19.44812  0.3137476            1 7.740898 0.34197044 0.0005344616
20  19.40982  0.2808348            1 8.964690 0.41974975 0.0021005286
     sigma_sq smooth_sigma_sq final_sigma_sq
1  0.15496968     0.007419036     0.15496968
2  0.02973391     0.007779416     0.02973391
3  0.02056017     0.005559263     0.02056017
4  0.02168590     0.005466293     0.02168590
5  5.98385838     0.675906363     5.98385838
6  0.03047738     0.008155760     0.03047738
7  4.97015558     0.436684492     4.97015558
8  3.22871046     0.979933557     3.22871046
9  0.02042512     0.005797009     0.02042512
10 0.04778499     0.011671534     0.04778499
11 5.66036598     1.065565643     5.66036598
12 0.02381314     0.008476591     0.02381314
13 0.04248844     0.012495401     0.04248844
14 0.01880838     0.006482768     0.01880838
15 6.71547409     0.174714511     6.71547409
16 7.07199652     0.086757951     7.07199652
17 0.02505069     0.005822397     0.02505069
18 0.02334533     0.008622043     0.02334533
19 0.03868398     0.006754404     0.03868398
20 0.03300382     0.005461667     0.03300382</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb134"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb134-1"><a href="#cb134-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save result table</span></span>
<span id="cb134-2"><a href="#cb134-2" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(sleuth_table_tx.il22.il26, <span class="at">file =</span> <span class="st">"./outputs/01_diffexp_sleuth-transcripts-qval005-IL22vsIL22+26.csv"</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Visualise these as a <strong>dittoSeq</strong> heatmap.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb135"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb135-1"><a href="#cb135-1" aria-hidden="true" tabindex="-1"></a>tx.diff<span class="fl">.2</span> <span class="ot">&lt;-</span> <span class="fu">na.omit</span>(sleuth_table_tx.il22.il26)</span>
<span id="cb135-2"><a href="#cb135-2" aria-hidden="true" tabindex="-1"></a>tx.diff<span class="fl">.2</span> <span class="ot">&lt;-</span> <span class="fu">paste</span>(tx.diff<span class="fl">.2</span><span class="sc">$</span>target_id, tx.diff<span class="fl">.2</span><span class="sc">$</span>ext_gene, <span class="at">sep =</span> <span class="st">"_"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb136"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb136-1"><a href="#cb136-1" aria-hidden="true" tabindex="-1"></a><span class="co"># By default, dittoHeatmap prioritises "logcounts" over "counts"</span></span>
<span id="cb136-2"><a href="#cb136-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dittoHeatmap</span>(bulkSCE, tx.diff<span class="fl">.2</span>,</span>
<span id="cb136-3"><a href="#cb136-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">annot.by =</span> <span class="fu">c</span>(<span class="st">"stim"</span>, <span class="st">"time"</span>, <span class="st">"medium"</span>, <span class="st">"donor"</span>), </span>
<span id="cb136-4"><a href="#cb136-4" aria-hidden="true" tabindex="-1"></a>             <span class="at">order.by =</span> (<span class="st">"stim"</span>), <span class="at">cluster_cols =</span> <span class="cn">FALSE</span>,</span>
<span id="cb136-5"><a href="#cb136-5" aria-hidden="true" tabindex="-1"></a>             <span class="at">main =</span> <span class="st">"Differential genes (qval&lt;0.05) | IL22+26 vs IL22"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="01_diffexp_sleuth_files/figure-html/unnamed-chunk-72-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="notes" class="level1">
<h1>Notes</h1>
<ul>
<li>Technical replicates for samples <code>A2-7*</code>, <code>A2-11*</code> and <code>A2-12*</code> are in progress.</li>
<li>Rebuild <strong>kallisto</strong> index with latest Ensembl release and re-run analysis.</li>
<li>All samples have also been mapped and counted using STAR with <code>--quantMode GeneCounts</code> for differential expression with <strong>DESeq2</strong> - are the main observations replicated?</li>
</ul>
</section>
<section id="packages-and-session-info" class="level1">
<h1>Packages and session info</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb137"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb137-1"><a href="#cb137-1" aria-hidden="true" tabindex="-1"></a>pkgs <span class="ot">&lt;-</span> <span class="fu">cite_packages</span>(<span class="at">output =</span> <span class="st">"table"</span>, <span class="at">out.dir =</span> <span class="st">"."</span>)</span>
<span id="cb137-2"><a href="#cb137-2" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(pkgs)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<thead>
<tr class="header">
<th style="text-align: left;">Package</th>
<th style="text-align: left;">Version</th>
<th style="text-align: left;">Citation</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">base</td>
<td style="text-align: left;">4.5.1</td>
<td style="text-align: left;"><span class="citation" data-cites="base">@base</span></td>
</tr>
<tr class="even">
<td style="text-align: left;">biomaRt</td>
<td style="text-align: left;">2.64.0</td>
<td style="text-align: left;"><span class="citation" data-cites="biomaRt">@biomaRt</span>….</td>
</tr>
<tr class="odd">
<td style="text-align: left;">dittoSeq</td>
<td style="text-align: left;">1.20.1</td>
<td style="text-align: left;"><span class="citation" data-cites="dittoSeq">@dittoSeq</span></td>
</tr>
<tr class="even">
<td style="text-align: left;">knitr</td>
<td style="text-align: left;">1.50</td>
<td style="text-align: left;"><span class="citation" data-cites="knitr20">@knitr20</span>….</td>
</tr>
<tr class="odd">
<td style="text-align: left;">PCAtools</td>
<td style="text-align: left;">2.20.0</td>
<td style="text-align: left;"><span class="citation" data-cites="PCAtools">@PCAtools</span></td>
</tr>
<tr class="even">
<td style="text-align: left;">rbioapi</td>
<td style="text-align: left;">0.8.3</td>
<td style="text-align: left;"><span class="citation" data-cites="rbioapi">@rbioapi</span></td>
</tr>
<tr class="odd">
<td style="text-align: left;">rmarkdown</td>
<td style="text-align: left;">2.30</td>
<td style="text-align: left;"><span class="citation" data-cites="rmarkdo">@rmarkdo</span>….</td>
</tr>
<tr class="even">
<td style="text-align: left;">sleuth</td>
<td style="text-align: left;">0.30.2</td>
<td style="text-align: left;"><span class="citation" data-cites="sleuth">@sleuth</span></td>
</tr>
<tr class="odd">
<td style="text-align: left;">SummarizedExperiment</td>
<td style="text-align: left;">1.38.1</td>
<td style="text-align: left;"><span class="citation" data-cites="Summari">@Summari</span>….</td>
</tr>
<tr class="even">
<td style="text-align: left;">tidyverse</td>
<td style="text-align: left;">2.0.0</td>
<td style="text-align: left;"><span class="citation" data-cites="tidyverse">@tidyverse</span></td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb138"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb138-1"><a href="#cb138-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.5.1 (2025-06-13)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0 
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_GB.UTF-8        LC_COLLATE=en_GB.UTF-8    
 [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8   
 [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       

time zone: Europe/London
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] grateful_0.3.0              rbioapi_0.8.3              
 [3] sleuth_0.30.2               PCAtools_2.20.0            
 [5] ggrepel_0.9.6               biomaRt_2.64.0             
 [7] dittoSeq_1.20.1             SummarizedExperiment_1.38.1
 [9] Biobase_2.68.0              GenomicRanges_1.60.0       
[11] GenomeInfoDb_1.44.3         IRanges_2.42.0             
[13] S4Vectors_0.46.0            BiocGenerics_0.54.1        
[15] generics_0.1.4              MatrixGenerics_1.20.0      
[17] matrixStats_1.5.0           lubridate_1.9.4            
[19] forcats_1.0.1               stringr_1.5.2              
[21] dplyr_1.1.4                 purrr_1.1.0                
[23] readr_2.1.5                 tidyr_1.3.1                
[25] tibble_3.3.0                ggplot2_4.0.0              
[27] tidyverse_2.0.0            

loaded via a namespace (and not attached):
 [1] DBI_1.2.3                   gridExtra_2.3              
 [3] httr2_1.2.1                 rlang_1.1.6                
 [5] magrittr_2.0.4              aggregation_1.0.1          
 [7] ggridges_0.5.7              compiler_4.5.1             
 [9] RSQLite_2.4.3               DelayedMatrixStats_1.30.0  
[11] reshape2_1.4.4              png_0.1-8                  
[13] vctrs_0.6.5                 pkgconfig_2.0.3            
[15] crayon_1.5.3                fastmap_1.2.0              
[17] dbplyr_2.5.1                XVector_0.48.0             
[19] labeling_0.4.3              rmarkdown_2.30             
[21] tzdb_0.5.0                  UCSC.utils_1.4.0           
[23] bit_4.6.0                   xfun_0.54                  
[25] cachem_1.1.0                beachmat_2.24.0            
[27] jsonlite_2.0.0              progress_1.2.3             
[29] blob_1.2.4                  rhdf5filters_1.20.0        
[31] DelayedArray_0.34.1         Rhdf5lib_1.30.0            
[33] BiocParallel_1.42.2         irlba_2.3.5.1              
[35] parallel_4.5.1              prettyunits_1.2.0          
[37] R6_2.6.1                    stringi_1.8.7              
[39] RColorBrewer_1.1-3          Rcpp_1.1.0                 
[41] knitr_1.50                  Matrix_1.7-4               
[43] timechange_0.3.0            tidyselect_1.2.1           
[45] rstudioapi_0.17.1           abind_1.4-8                
[47] yaml_2.3.10                 codetools_0.2-19           
[49] curl_7.0.0                  plyr_1.8.9                 
[51] lattice_0.22-5              withr_3.0.2                
[53] KEGGREST_1.48.1             S7_0.2.0                   
[55] evaluate_1.0.5              BiocFileCache_2.16.2       
[57] xml2_1.4.1                  Biostrings_2.76.0          
[59] pillar_1.11.1               filelock_1.0.3             
[61] renv_1.1.5                  hms_1.1.4                  
[63] sparseMatrixStats_1.20.0    scales_1.4.0               
[65] glue_1.8.0                  lazyeval_0.2.2             
[67] pheatmap_1.0.13             tools_4.5.1                
[69] data.table_1.17.8           ScaledMatrix_1.16.0        
[71] rhdf5_2.52.1                cowplot_1.2.0              
[73] grid_4.5.1                  AnnotationDbi_1.70.0       
[75] colorspace_2.1-2            SingleCellExperiment_1.30.1
[77] GenomeInfoDbData_1.2.14     BiocSingular_1.24.0        
[79] cli_3.6.5                   rsvd_1.0.5                 
[81] rappdirs_0.3.3              S4Arrays_1.8.1             
[83] gtable_0.3.6                digest_0.6.37              
[85] dqrng_0.4.1                 SparseArray_1.8.1          
[87] farver_2.1.2                memoise_2.0.1              
[89] htmltools_0.5.8.1           lifecycle_1.0.4            
[91] httr_1.4.7                  bit64_4.6.0-1              </code></pre>
</div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>