<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>3. Biological themes classification and the enrichment analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="03_clusterprofiler_files/libs/clipboard/clipboard.min.js"></script>
<script src="03_clusterprofiler_files/libs/quarto-html/quarto.js"></script>
<script src="03_clusterprofiler_files/libs/quarto-html/popper.min.js"></script>
<script src="03_clusterprofiler_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="03_clusterprofiler_files/libs/quarto-html/anchor.min.js"></script>
<link href="03_clusterprofiler_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="03_clusterprofiler_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="03_clusterprofiler_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="03_clusterprofiler_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="03_clusterprofiler_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">3. Biological themes classification and the enrichment analysis</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<p>V Chong | vanessa.chong@rdm.ox.ac.uk | 16 November 2025</p>
<section id="preamble" class="level1">
<h1>Preamble</h1>
<ul>
<li>Bulk RNAseq samples from healthy, patient-derived organoids of the terminal ileum (TI).</li>
<li>Libraries were prepared and sequenced in several batches.</li>
<li>Organoids were stimulated with IL22 only, both IL22 and IL26, or IL26 only. Controls were unstimulated.</li>
<li>Stimulation was performed for 24, 48, or 72 hours.</li>
</ul>
<section id="points-of-note" class="level2">
<h2 class="anchored" data-anchor-id="points-of-note">Points-of-note</h2>
<ul>
<li>24 hours samples are spread across three batches, and two different (previous and current) growth mediums. <strong>technical replicates in progress</strong></li>
<li>48 hours samples constitute a batch of its own, performed in current growth medium only. <strong>technical replicates in progress</strong></li>
<li>72 hours samples are not replicated and performed in a previous growth medium only - these were excluded in previous differential expression analyses.</li>
</ul>
</section>
</section>
<section id="analysis" class="level1">
<h1>Analysis</h1>
<ul>
<li><p>Previous pairwise differential expression between each IL stimulation vs control followed by Venn intersections of differential genes indicated there could be an IL26-mediated effect on IL22 stimulation. IL26 on its own did not appear to drive significant effect(s).</p></li>
<li><p><a href="https://doi.org/10.1089/omi.2011.01"><strong>clusterProfiler</strong></a> biological themes classification and the enrichment analysis to compare effects between IL22 stimulation and IL22+IL26 co-stimulation.</p></li>
</ul>
<section id="excerpt-from-the-clusterprofiler-cookbook" class="level2">
<h2 class="anchored" data-anchor-id="excerpt-from-the-clusterprofiler-cookbook">Excerpt from the <a href="https://yulab-smu.top/biomedical-knowledge-mining-book/index.html"><strong>clusterProfiler</strong> cookbook</a>:</h2>
<ul>
<li><p><em>Over Representation Analysis (ORA) (Boyle et al.&nbsp;2004) is a widely used approach to determine whether known biological functions or processes are over-represented (= enriched) in an experimentally-derived gene list, e.g.&nbsp;a list of differentially expressed genes (DEGs).</em></p></li>
<li><p><em>A common approach to analyzing gene expression profiles is identifying differentially expressed genes that are deemed interesting. The ORA enrichment analysis is based on these differentially expressed genes. This approach will find genes where the difference is large and will fail where the difference is small, but evidenced in coordinated way in a set of related genes. Gene Set Enrichment Analysis (GSEA) (Subramanian et al.&nbsp;2005) directly addresses this limitation. All genes can be used in GSEA; GSEA aggregates the per gene statistics across genes within a gene set, therefore making it possible to detect situations where all genes in a predefined set change in a small but coordinated way. This is important since it is likely that many relevant phenotypic differences are manifested by small but consistent changes in a set of genes.</em></p></li>
</ul>
</section>
<section id="initialise-r-environment" class="level2">
<h2 class="anchored" data-anchor-id="initialise-r-environment">Initialise R environment</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="do">##### Set libPaths #####</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">.libPaths</span>(<span class="fu">c</span>(<span class="st">"/home/chongmorrison/R/x86_64-pc-linux-gnu-library/4.5"</span>,</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>            <span class="st">"/home/chongmorrison/R/4.5.0-Bioc3.21"</span>))</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">.libPaths</span>() <span class="co"># check .libPaths</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "/home/chongmorrison/R/x86_64-pc-linux-gnu-library/4.5"
[2] "/home/chongmorrison/R/4.5.0-Bioc3.21"                 
[3] "/usr/local/lib/R/site-library"                        
[4] "/usr/lib/R/site-library"                              
[5] "/usr/lib/R/library"                                   </code></pre>
</div>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="do">##### Memory/parallel cores usage #####</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">Ncpus =</span> <span class="dv">12</span>) <span class="co"># adjust no. of cores for base R</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">getOption</span>(<span class="st">"Ncpus"</span>, <span class="dv">1</span>L)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 12</code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="do">##### Import all required packages #####</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="co"># data-wrangling and plots</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(viridis)</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="co">#library(ggpubr)</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SummarizedExperiment)</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dittoSeq)</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a><span class="co"># analysis</span></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(sleuth)</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(clusterProfiler)</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(enrichplot)</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rbioapi)</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a><span class="co"># to cite packages</span></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(grateful)</span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a><span class="do">##### Set seed for reproducibility of random sampling #####</span></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">584</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="import-differential-expression-gene-lists" class="level2">
<h2 class="anchored" data-anchor-id="import-differential-expression-gene-lists">Import differential expression gene lists</h2>
<p>To get as “clean” results as possible for downstream analyses, the same parameters were employed for all differential expression:</p>
<ul>
<li><code>sleuth_lrt</code> Likelihood Ratio test, accounting for <code>batch</code> and <code>donor</code> differences</li>
<li><code>sleuth_results</code> with <code>pval_aggregate = FALSE</code> to extract transcript-level results</li>
<li><code>qval &lt;= 0.05</code> filter for differentially-expressed transcripts</li>
</ul>
<p><strong>IL22 only vs control</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>il22.ctl <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="at">file =</span> <span class="st">"./outputs/02_diffexp_sleuth-transcripts-qval005_IL22-ctl.csv"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>IL26 only vs control</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>il26.ctl <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="at">file =</span> <span class="st">"./outputs/02_diffexp_sleuth-transcripts-qval005_IL26-ctl.csv"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>IL22+26 vs control</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>il26.<span class="fl">22.</span>ctl <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(<span class="at">file =</span> <span class="st">"./outputs/02_diffexp_sleuth-transcripts-qval005_IL22+26-ctl.csv"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="import-background-genes" class="level3">
<h3 class="anchored" data-anchor-id="import-background-genes">Import background genes</h3>
<p>This should be all the genes used in the differential expression analysis.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>so <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"./02_diffexp_sleuth-so_72h-excluded.rds"</span>)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>ttg <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"./ttg115.rds"</span>)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>abundances <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(so[[<span class="st">"obs_raw"</span>]])</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>abundances.gene <span class="ot">&lt;-</span> <span class="fu">left_join</span>(abundances, ttg, <span class="at">by =</span> <span class="st">"target_id"</span>)</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>bg.genes <span class="ot">&lt;-</span> <span class="fu">unique</span>(abundances.gene<span class="sc">$</span>ext_gene)</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(bg.genes) <span class="co"># 30368 genes</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 30368</code></pre>
</div>
</div>
<p>Convert to Entrez ID using <code>bitr()</code> utility included in <strong>clusterProfiler</strong>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>bg.genes <span class="ot">&lt;-</span> <span class="fu">bitr</span>(bg.genes, <span class="at">fromType =</span> <span class="st">"SYMBOL"</span>, <span class="at">toType =</span> <span class="st">"ENTREZID"</span>, <span class="at">OrgDb =</span> <span class="st">"org.Hs.eg.db"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code></code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>'select()' returned 1:many mapping between keys and columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in bitr(bg.genes, fromType = "SYMBOL", toType = "ENTREZID", OrgDb =
"org.Hs.eg.db"): 0.08% of input gene IDs are fail to map...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(bg.genes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   SYMBOL ENTREZID
1    ARF5      381
2    M6PR     4074
3   ESRRA     2101
4   FKBP4     2288
5 CYP26B1    56603
6 NDUFAF7    55471</code></pre>
</div>
</div>
</section>
<section id="import-tpm-values" class="level3">
<h3 class="anchored" data-anchor-id="import-tpm-values">Import TPM values</h3>
<p>This is to compute “fold changes”.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Chain `target_id` (ensembl transcript ID with version) with `ext_gene` (gene symbol)</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>abundances.gene<span class="sc">$</span>target_id.gene <span class="ot">&lt;-</span> <span class="fu">paste</span>(abundances.gene<span class="sc">$</span>target_id, abundances.gene<span class="sc">$</span>ext_gene, <span class="at">sep =</span> <span class="st">"_"</span>)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="co"># transcript-per-million values</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>tpm <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(abundances.gene<span class="sc">$</span>target_id.gene, abundances.gene<span class="sc">$</span>tpm, abundances.gene<span class="sc">$</span>sample) <span class="sc">%&gt;%</span> <span class="fu">setNames</span>(<span class="fu">c</span>(<span class="st">"target_id"</span>, <span class="st">"tpm"</span>, <span class="st">"sample"</span>))</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>tpm <span class="ot">&lt;-</span> tpm <span class="sc">%&gt;%</span> <span class="fu">pivot_wider</span>(<span class="at">names_from=</span>sample, <span class="at">values_from=</span>tpm) <span class="sc">%&gt;%</span> <span class="fu">data.frame</span>(<span class="at">check.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(tpm) <span class="ot">&lt;-</span> tpm<span class="sc">$</span>target_id</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>tpm <span class="ot">&lt;-</span> <span class="fu">subset</span>(tpm, <span class="at">select=</span><span class="sc">-</span><span class="fu">c</span>(target_id))</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(tpm[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                             A2-7-A1   A2-7-A2     A2-7-A3   A2-7-A4
ENST00000000233.10_ARF5   55.4104139 85.040130 108.2232382 89.359572
ENST00000000412.8_M6PR    45.7325370 40.418322  45.9905393 49.847565
ENST00000000442.11_ESRRA  33.1923120 39.944588  54.3414653 43.194645
ENST00000001008.6_FKBP4    0.4535558  0.751447   0.7904457  0.456891
ENST00000001146.7_CYP26B1 31.7898330 21.999630  37.4672535 24.110882
                               A2-7-B1
ENST00000000233.10_ARF5    93.40026757
ENST00000000412.8_M6PR     37.18590347
ENST00000000442.11_ESRRA   83.22098866
ENST00000001008.6_FKBP4     0.02386233
ENST00000001146.7_CYP26B1 116.38585933</code></pre>
</div>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># log(tpm+1) values</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>logtpm <span class="ot">&lt;-</span> <span class="fu">log2</span>(tpm <span class="sc">+</span> <span class="dv">1</span>)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(logtpm[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                            A2-7-A1   A2-7-A2   A2-7-A3  A2-7-A4    A2-7-B1
ENST00000000233.10_ARF5   5.8178896 6.4269378 6.7711360 6.497606 6.56071904
ENST00000000412.8_M6PR    5.5463555 5.3721972 5.5542984 5.668107 5.25496825
ENST00000000442.11_ESRRA  5.0956001 5.3556009 5.7902889 5.465800 6.39610791
ENST00000001008.6_FKBP4   0.5395865 0.8085473 0.8403187 0.542893 0.03402174
ENST00000001146.7_CYP26B1 5.0351766 4.5235387 5.2655589 4.650241 6.87511482</code></pre>
</div>
</div>
</section>
</section>
<section id="over-representation-analysis" class="level2">
<h2 class="anchored" data-anchor-id="over-representation-analysis">Over-representation analysis</h2>
<section id="kegg-kyoto-encyclopedia-of-genes-and-genomes" class="level3">
<h3 class="anchored" data-anchor-id="kegg-kyoto-encyclopedia-of-genes-and-genomes"><a href="https://www.genome.jp/kegg/">KEGG: Kyoto Encyclopedia of Genes and Genomes</a></h3>
<p><strong>IL26 only vs control</strong></p>
<p>IL26 only vs control differential expression resulted in only 14 transcripts at <code>qval &lt;= 0.05</code> i.e.&nbsp;implicating that IL26 stimulation alone has minimal biological effect. Over-representation of pathway(s) very unlikely:</p>
<p>Obtain list of differential genes <code>qval &lt;= 0.05</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>il26.genes <span class="ot">&lt;-</span> <span class="fu">unique</span>(il26.ctl<span class="sc">$</span>ext_gene)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(il26.genes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 9</code></pre>
</div>
</div>
<p>Convert to Entrez ID using <code>bitr()</code> utility included in <strong>clusterProfiler</strong>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>il26.genes <span class="ot">&lt;-</span> <span class="fu">bitr</span>(il26.genes, <span class="at">fromType =</span> <span class="st">"SYMBOL"</span>, <span class="at">toType =</span> <span class="st">"ENTREZID"</span>, <span class="at">OrgDb =</span> <span class="st">"org.Hs.eg.db"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>'select()' returned 1:1 mapping between keys and columns</code></pre>
</div>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(il26.genes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   SYMBOL ENTREZID
1   DMBT1     1755
2 SMARCB1     6598
3   RBCK1    10616
4    MUC1     4582
5   KDM5C     8242
6 TBC1D17    79735</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>il26.kegg.ora <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">enrichKEGG</span>(<span class="at">gene =</span> il26.genes<span class="sc">$</span>ENTREZID, <span class="at">organism =</span> <span class="st">"hsa"</span>, <span class="at">universe =</span> bg.genes<span class="sc">$</span>ENTREZID,</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>                                       <span class="at">pvalueCutoff =</span> <span class="fl">0.05</span>, <span class="at">pAdjustMethod =</span> <span class="st">"BH"</span>, <span class="at">qvalueCutoff =</span> <span class="fl">0.2</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Reading KEGG annotation online: "https://rest.kegg.jp/link/hsa/pathway"...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Reading KEGG annotation online: "https://rest.kegg.jp/list/pathway/hsa"...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(il26.kegg.ora, <span class="dv">10</span>) <span class="co"># no categories over-represented</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] category       subcategory    ID             Description    GeneRatio     
 [6] BgRatio        RichFactor     FoldEnrichment zScore         pvalue        
[11] p.adjust       qvalue         geneID         Count         
&lt;0 rows&gt; (or 0-length row.names)</code></pre>
</div>
</div>
<p><strong>IL22 only vs control</strong></p>
<p>IL22 stimulation resulted in 4868 differential transcripts at <code>qval &lt;= 0.05</code>. Filter to top ~500-1000 hits with <code>qval &lt;= 0.001</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>il22.genes <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(il22.ctl, qval <span class="sc">&lt;=</span> <span class="fl">0.001</span>)</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>il22.genes <span class="ot">&lt;-</span> <span class="fu">unique</span>(il22.genes<span class="sc">$</span>ext_gene)</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(il22.genes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 881</code></pre>
</div>
</div>
<p>Convert to Entrez ID using <code>bitr()</code> utility with <strong>clusterProfiler</strong>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>il22.genes <span class="ot">&lt;-</span> <span class="fu">bitr</span>(il22.genes, <span class="at">fromType =</span> <span class="st">"SYMBOL"</span>, <span class="at">toType =</span> <span class="st">"ENTREZID"</span>, <span class="at">OrgDb =</span> <span class="st">"org.Hs.eg.db"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>'select()' returned 1:1 mapping between keys and columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in bitr(il22.genes, fromType = "SYMBOL", toType = "ENTREZID", OrgDb =
"org.Hs.eg.db"): 0.11% of input gene IDs are fail to map...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(il22.genes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  SYMBOL ENTREZID
1  DMBT1     1755
2 IFITM1     8519
3 UNC5CL   222643
4 SPINK1     6690
5  REG1A     5967
6    CFB      629</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>il22.kegg.ora <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">enrichKEGG</span>(<span class="at">gene =</span> il22.genes<span class="sc">$</span>ENTREZID, <span class="at">organism =</span> <span class="st">"hsa"</span>, <span class="at">universe =</span> bg.genes<span class="sc">$</span>ENTREZID,</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>                                       <span class="at">pvalueCutoff =</span> <span class="fl">0.05</span>, <span class="at">pAdjustMethod =</span> <span class="st">"BH"</span>, <span class="at">qvalueCutoff =</span> <span class="fl">0.2</span>))</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(il22.kegg.ora, <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                               category
hsa05204                 Human Diseases
hsa00051                     Metabolism
hsa00980                     Metabolism
hsa00982                     Metabolism
hsa04141 Genetic Information Processing
hsa00500                     Metabolism
hsa00590                     Metabolism
hsa00830                     Metabolism
hsa01200                     Metabolism
hsa00140                     Metabolism
                                       subcategory       ID
hsa05204                          Cancer: overview hsa05204
hsa00051                   Carbohydrate metabolism hsa00051
hsa00980 Xenobiotics biodegradation and metabolism hsa00980
hsa00982 Xenobiotics biodegradation and metabolism hsa00982
hsa04141          Folding, sorting and degradation hsa04141
hsa00500                   Carbohydrate metabolism hsa00500
hsa00590                          Lipid metabolism hsa00590
hsa00830      Metabolism of cofactors and vitamins hsa00830
hsa01200                  Global and overview maps hsa01200
hsa00140                          Lipid metabolism hsa00140
                                          Description GeneRatio  BgRatio
hsa05204        Chemical carcinogenesis - DNA adducts    19/504  70/8334
hsa00051              Fructose and mannose metabolism    13/504  35/8334
hsa00980 Metabolism of xenobiotics by cytochrome P450    19/504  77/8334
hsa00982            Drug metabolism - cytochrome P450    16/504  71/8334
hsa04141  Protein processing in endoplasmic reticulum    25/504 170/8334
hsa00500                Starch and sucrose metabolism    10/504  36/8334
hsa00590                  Arachidonic acid metabolism    13/504  62/8334
hsa00830                           Retinol metabolism    13/504  68/8334
hsa01200                            Carbon metabolism    18/504 116/8334
hsa00140                 Steroid hormone biosynthesis    12/504  64/8334
         RichFactor FoldEnrichment   zScore       pvalue     p.adjust
hsa05204  0.2714286       4.488265 7.435299 1.783470e-08 5.938955e-06
hsa00051  0.3714286       6.141837 7.733472 5.354791e-08 8.915727e-06
hsa00980  0.2467532       4.080241 6.888966 9.583002e-08 1.063713e-05
hsa00982  0.2253521       3.726358 5.852997 3.630540e-06 3.022424e-04
hsa04141  0.1470588       2.431723 4.784835 2.912017e-05 1.939403e-03
hsa00500  0.2777778       4.593254 5.481350 3.640941e-05 2.020722e-03
hsa00590  0.2096774       3.467166 4.946796 6.664017e-05 3.170168e-03
hsa00830  0.1911765       3.161239 4.539884 1.802272e-04 7.034659e-03
hsa01200  0.1551724       2.565887 4.308651 1.901259e-04 7.034659e-03
hsa00140  0.1875000       3.100446 4.279410 3.832679e-04 1.123690e-02
               qvalue
hsa05204 5.237769e-06
hsa00051 7.863088e-06
hsa00980 9.381255e-06
hsa00982 2.665580e-04
hsa04141 1.710427e-03
hsa00500 1.782145e-03
hsa00590 2.795881e-03
hsa00830 6.204109e-03
hsa01200 6.204109e-03
hsa00140 9.910215e-03
                                                                                                                                      geneID
hsa05204                                       79799/1551/2939/1576/6822/2938/1562/54576/1543/2946/10/1559/6799/7364/2948/1577/4259/873/1557
hsa00051                                                                 3099/57016/26007/7264/5373/3098/6652/5209/2203/2762/3795/29925/5208
hsa00980                                    1645/29785/79799/2939/1576/6822/2938/54576/1543/22977/2946/127/1559/7364/2948/1577/4259/873/8574
hsa00982                                                    4128/79799/2939/1576/2938/4129/54576/2330/2946/127/1559/7364/2948/1577/4259/1557
hsa04141 10525/7494/51726/9601/4121/2923/6747/58505/3309/27248/5611/23480/811/10130/3337/10952/6745/3998/30001/9871/573/6184/1650/29927/9695
hsa00500                                                                               3099/8972/6476/5169/57733/3098/283209/2538/5834/11181
hsa00590                                                                5320/54979/11145/1573/2879/1559/80142/3248/8529/113612/873/2053/1557
hsa00830                                                                   8694/29785/79799/1551/1576/1562/54576/1543/127/216/1559/7364/1577
hsa01200                                            3099/26007/1373/5230/84532/50/4967/3098/132158/2023/2203/5313/847/35/9104/4329/3417/8802
hsa00140                                                                       3294/1645/79799/1551/1576/54576/1543/79154/412/3283/7364/1577
         Count
hsa05204    19
hsa00051    13
hsa00980    19
hsa00982    16
hsa04141    25
hsa00500    10
hsa00590    13
hsa00830    13
hsa01200    18
hsa00140    12</code></pre>
</div>
</div>
<p><strong>IL22+26 vs control</strong></p>
<p>IL22+26 stimulation resulted in 4584 differential transcripts at <code>qval &lt;= 0.05</code>. Filter to top ~500-1000 hits with <code>qval &lt;= 0.001</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>il26.<span class="fl">22.</span>genes <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(il26.<span class="fl">22.</span>ctl, qval <span class="sc">&lt;=</span> <span class="fl">0.001</span>)</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>il26.<span class="fl">22.</span>genes <span class="ot">&lt;-</span> <span class="fu">unique</span>(il26.<span class="fl">22.</span>genes<span class="sc">$</span>ext_gene)</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(il26.<span class="fl">22.</span>genes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 857</code></pre>
</div>
</div>
<p>Convert to Entrez ID using <code>bitr()</code> utility with <strong>clusterProfiler</strong>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>il26.<span class="fl">22.</span>genes <span class="ot">&lt;-</span> <span class="fu">bitr</span>(il26.<span class="fl">22.</span>genes, <span class="at">fromType =</span> <span class="st">"SYMBOL"</span>, <span class="at">toType =</span> <span class="st">"ENTREZID"</span>, <span class="at">OrgDb =</span> <span class="st">"org.Hs.eg.db"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>'select()' returned 1:1 mapping between keys and columns</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in bitr(il26.22.genes, fromType = "SYMBOL", toType = "ENTREZID", :
0.12% of input gene IDs are fail to map...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(il26.<span class="fl">22.</span>genes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    SYMBOL ENTREZID
1    DMBT1     1755
2   CYBRD1    79901
3   UNC5CL   222643
4      OTC     5009
5 SERPINA3       12
6   SPINK1     6690</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>il26.<span class="fl">22.</span>kegg.ora <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">enrichKEGG</span>(<span class="at">gene =</span> il26.<span class="fl">22.</span>genes<span class="sc">$</span>ENTREZID, <span class="at">organism =</span> <span class="st">"hsa"</span>, <span class="at">universe =</span> bg.genes<span class="sc">$</span>ENTREZID,</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>                                          <span class="at">pvalueCutoff =</span> <span class="fl">0.05</span>, <span class="at">pAdjustMethod =</span> <span class="st">"BH"</span>, <span class="at">qvalueCutoff =</span> <span class="fl">0.2</span>))</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(il26.<span class="fl">22.</span>kegg.ora, <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                   category                               subcategory       ID
hsa00980         Metabolism Xenobiotics biodegradation and metabolism hsa00980
hsa05204     Human Diseases                          Cancer: overview hsa05204
hsa00590         Metabolism                          Lipid metabolism hsa00590
hsa00982         Metabolism Xenobiotics biodegradation and metabolism hsa00982
hsa00591         Metabolism                          Lipid metabolism hsa00591
hsa00830         Metabolism      Metabolism of cofactors and vitamins hsa00830
hsa00140         Metabolism                          Lipid metabolism hsa00140
hsa04146 Cellular Processes                  Transport and catabolism hsa04146
hsa00051         Metabolism                   Carbohydrate metabolism hsa00051
hsa00500         Metabolism                   Carbohydrate metabolism hsa00500
                                          Description GeneRatio BgRatio
hsa00980 Metabolism of xenobiotics by cytochrome P450    22/487 77/8334
hsa05204        Chemical carcinogenesis - DNA adducts    20/487 70/8334
hsa00590                  Arachidonic acid metabolism    16/487 62/8334
hsa00982            Drug metabolism - cytochrome P450    17/487 71/8334
hsa00591                     Linoleic acid metabolism    11/487 30/8334
hsa00830                           Retinol metabolism    15/487 68/8334
hsa00140                 Steroid hormone biosynthesis    14/487 64/8334
hsa04146                                   Peroxisome    16/487 83/8334
hsa00051              Fructose and mannose metabolism    10/487 35/8334
hsa00500                Starch and sucrose metabolism    10/487 36/8334
         RichFactor FoldEnrichment   zScore       pvalue     p.adjust
hsa00980  0.2857143       4.889410 8.541445 2.310490e-10 7.485988e-08
hsa05204  0.2857143       4.889410 8.140498 1.540907e-09 2.496270e-07
hsa00590  0.2580645       4.416242 6.725933 3.236513e-07 3.109887e-05
hsa00982  0.2394366       4.097464 6.529499 4.463732e-07 3.109887e-05
hsa00591  0.3666667       6.274743 7.209940 4.799209e-07 3.109887e-05
hsa00830  0.2205882       3.774912 5.723601 6.374701e-06 3.442338e-04
hsa00140  0.2187500       3.743455 5.488433 1.436529e-05 6.649076e-04
hsa04146  0.1927711       3.298879 5.243427 1.953892e-05 7.424505e-04
hsa00051  0.2857143       4.889410 5.744051 2.062362e-05 7.424505e-04
hsa00500  0.2777778       4.753593 5.622444 2.708293e-05 8.774870e-04
               qvalue
hsa00980 6.323447e-08
hsa05204 2.108610e-07
hsa00590 2.626935e-05
hsa00982 2.626935e-05
hsa00591 2.626935e-05
hsa00830 2.907758e-04
hsa00140 5.616503e-04
hsa04146 6.271512e-04
hsa00051 6.271512e-04
hsa00500 7.412171e-04
                                                                                                                    geneID
hsa00980 29785/1645/79799/1577/2938/2939/54576/1576/22977/221/2946/1543/2948/126/8574/873/246181/1559/7364/4259/6822/54575
hsa05204               79799/1577/1557/2938/1562/2939/54576/1576/1551/2946/1543/2948/10/873/1559/7364/4259/6822/6799/54575
hsa00590                            1557/5320/54979/1573/2879/11145/2053/873/1559/5321/151056/8398/4051/113612/80142/84647
hsa00982                             79799/1577/1557/2938/2939/54576/4128/1576/221/2946/2948/126/1559/7364/2330/4259/54575
hsa00591                                                       1557/5320/1576/54979/1573/11145/1559/5321/151056/8398/84647
hsa00830                                      29785/79799/1577/1562/54576/1576/1551/1543/8694/126/5959/1559/7364/216/54575
hsa00140                                         3294/1645/79799/1577/54576/79154/1576/51170/1551/1543/412/6783/7364/54575
hsa04146                                54677/4843/5264/8309/11001/6647/7498/51703/8800/3417/6342/2053/847/23600/5827/1384
hsa00051                                                               57016/26007/3099/2203/3098/5208/6652/5373/7264/5209
hsa00500                                                             6476/3099/57733/283209/8972/2538/3098/11181/5169/5834
         Count
hsa00980    22
hsa05204    20
hsa00590    16
hsa00982    17
hsa00591    11
hsa00830    15
hsa00140    14
hsa04146    16
hsa00051    10
hsa00500    10</code></pre>
</div>
</div>
<section id="compare-multiple-gene-lists" class="level4">
<h4 class="anchored" data-anchor-id="compare-multiple-gene-lists">Compare multiple gene lists</h4>
<p>Compile list of gene lists.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>compare.list <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">il22 =</span> il22.genes<span class="sc">$</span>ENTREZID,</span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>                     <span class="at">il26.22 =</span> il26.<span class="fl">22.</span>genes<span class="sc">$</span>ENTREZID)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>compare.kegg <span class="ot">&lt;-</span> <span class="fu">compareCluster</span>(<span class="at">geneCluster =</span> compare.list, <span class="at">fun =</span> <span class="st">"enrichKEGG"</span>, <span class="at">universe =</span> bg.genes<span class="sc">$</span>ENTREZID,</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>                               <span class="at">organism =</span> <span class="st">"hsa"</span>, <span class="at">pvalueCutoff =</span> <span class="fl">0.05</span>, <span class="at">pAdjustMethod =</span> <span class="st">"BH"</span>, <span class="at">qvalueCutoff =</span> <span class="fl">0.2</span>)</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(compare.kegg)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Cluster                       category
1    il22                 Human Diseases
2    il22                     Metabolism
3    il22                     Metabolism
4    il22                     Metabolism
5    il22 Genetic Information Processing
6    il22                     Metabolism
                                subcategory       ID
1                          Cancer: overview hsa05204
2                   Carbohydrate metabolism hsa00051
3 Xenobiotics biodegradation and metabolism hsa00980
4 Xenobiotics biodegradation and metabolism hsa00982
5          Folding, sorting and degradation hsa04141
6                   Carbohydrate metabolism hsa00500
                                   Description GeneRatio  BgRatio RichFactor
1        Chemical carcinogenesis - DNA adducts    19/504  70/8334  0.2714286
2              Fructose and mannose metabolism    13/504  35/8334  0.3714286
3 Metabolism of xenobiotics by cytochrome P450    19/504  77/8334  0.2467532
4            Drug metabolism - cytochrome P450    16/504  71/8334  0.2253521
5  Protein processing in endoplasmic reticulum    25/504 170/8334  0.1470588
6                Starch and sucrose metabolism    10/504  36/8334  0.2777778
  FoldEnrichment   zScore       pvalue     p.adjust       qvalue
1       4.488265 7.435299 1.783470e-08 5.938955e-06 5.237769e-06
2       6.141837 7.733472 5.354791e-08 8.915727e-06 7.863088e-06
3       4.080241 6.888966 9.583002e-08 1.063713e-05 9.381255e-06
4       3.726358 5.852997 3.630540e-06 3.022424e-04 2.665580e-04
5       2.431723 4.784835 2.912017e-05 1.939403e-03 1.710427e-03
6       4.593254 5.481350 3.640941e-05 2.020722e-03 1.782145e-03
                                                                                                                               geneID
1                                       79799/1551/2939/1576/6822/2938/1562/54576/1543/2946/10/1559/6799/7364/2948/1577/4259/873/1557
2                                                                 3099/57016/26007/7264/5373/3098/6652/5209/2203/2762/3795/29925/5208
3                                    1645/29785/79799/2939/1576/6822/2938/54576/1543/22977/2946/127/1559/7364/2948/1577/4259/873/8574
4                                                    4128/79799/2939/1576/2938/4129/54576/2330/2946/127/1559/7364/2948/1577/4259/1557
5 10525/7494/51726/9601/4121/2923/6747/58505/3309/27248/5611/23480/811/10130/3337/10952/6745/3998/30001/9871/573/6184/1650/29927/9695
6                                                                               3099/8972/6476/5169/57733/3098/283209/2538/5834/11181
  Count
1    19
2    13
3    19
4    16
5    25
6    10</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="co"># clusterProfiler dotplot</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dotplot</span>(compare.kegg, <span class="at">showCategory =</span> <span class="dv">100</span>,</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">x =</span> <span class="st">"Cluster"</span>, <span class="at">color =</span> <span class="st">"FoldEnrichment"</span>, <span class="at">size =</span> <span class="st">"geneRatio"</span>,</span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">font.size =</span> <span class="dv">5</span>, <span class="at">title =</span> <span class="st">"KEGG over-representation | diff genes with qval &lt;= 0.001"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: `aes_string()` was deprecated in ggplot2 3.0.0.
ℹ Please use tidy evaluation idioms with `aes()`.
ℹ See also `vignette("ggplot2-in-packages")` for more information.
ℹ The deprecated feature was likely used in the enrichplot package.
  Please report the issue at
  &lt;https://github.com/GuangchuangYu/enrichplot/issues&gt;.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="03_clusterprofiler_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get data.frame of output</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>compare.kegg_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(compare.kegg)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="gene-ontology-go-consortium" class="level3">
<h3 class="anchored" data-anchor-id="gene-ontology-go-consortium"><a href="https://geneontology.org/">Gene Ontology (GO) Consortium</a></h3>
<p>GO comprises three orthogonal ontologies i.e.&nbsp;molecular function (MF), biological process (BP), and cellular component (CC).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>compare.go <span class="ot">&lt;-</span> <span class="fu">compareCluster</span>(<span class="at">geneCluster =</span> compare.list, <span class="at">fun =</span> <span class="st">"enrichGO"</span>, <span class="at">ont =</span> <span class="st">"BP"</span>, <span class="at">universe =</span> bg.genes<span class="sc">$</span>ENTREZID,</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>                             <span class="at">OrgDb =</span> <span class="st">"org.Hs.eg.db"</span>, <span class="at">pvalueCutoff =</span> <span class="fl">0.05</span>, <span class="at">pAdjustMethod =</span> <span class="st">"BH"</span>, <span class="at">qvalueCutoff =</span> <span class="fl">0.2</span>,</span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>                             <span class="at">readable =</span> <span class="cn">TRUE</span>)</span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(compare.go)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Cluster         ID                              Description GeneRatio
1    il22 GO:0006805             xenobiotic metabolic process    31/825
2    il22 GO:0044282         small molecule catabolic process    51/825
3    il22 GO:0006631             fatty acid metabolic process    53/825
4    il22 GO:0009410          response to xenobiotic stimulus    55/825
5    il22 GO:0071466 cellular response to xenobiotic stimulus    34/825
6    il22 GO:0006066                alcohol metabolic process    48/825
    BgRatio RichFactor FoldEnrichment    zScore       pvalue     p.adjust
1 126/17563  0.2460317       5.237643 10.598294 1.533238e-14 7.526664e-11
2 366/17563  0.1393443       2.966428  8.440234 2.970952e-12 7.292202e-09
3 402/17563  0.1318408       2.806691  8.135565 9.709956e-12 1.588872e-08
4 440/17563  0.1250000       2.661061  7.834009 3.214816e-11 3.945383e-08
5 198/17563  0.1717172       3.655598  8.342953 4.696197e-11 4.610727e-08
6 361/17563  0.1329640       2.830602  7.802256 7.020497e-11 5.743937e-08
        qvalue
1 6.233015e-11
2 6.038851e-09
3 1.315784e-08
4 3.267268e-08
5 3.818256e-08
6 4.756695e-08
                                                                                                                                                                                                                                                                                                                                              geneID
1                                                                                                                                       AKR1C1/ABCC2/BPHL/CYP2S1/AADAC/CYP3A7/GSTA2/CYP3A4/SULT2A1/GSTA1/CYP2C18/HNF4A/ALDH2/UGT1A8/CES2/FMO5/CYP1A1/SLC28A2/GSTM2/CYP2J2/NAT2/CYP2C9/SULT1A2/CYP4F2/CYP2U1/UGT2B7/GSTM4/CYP3A5/CBR1/CYP4F12/CYP2C19
2                                               IRS1/GDA/OTC/HK2/AKR1B10/CROT/ABAT/XDH/NOS2/TKFC/ALDH5A1/AMACR/PCK1/CYP3A4/ABHD2/SULT2A1/OAT/PGK1/PHYH/PMM2/ACSS1/HK1/SLC27A2/ALDH2/ATP2B4/RBKS/GLYCTK/BCKDHB/CYP1A1/ADTRP/ACOX2/PCK2/ADH4/SORD/ENO1/QPRT/ASRGL1/SULT1A2/IRS2/CYP4F2/PNKD/ARG2/ACADS/PDXP/GK5/ALDH6A1/HAAO/MCCC2/PFKFB2/FAAH/CYP4F12
3                   IRS1/AKR1C1/CROT/XBP1/CYP2S1/UBR4/CEACAM1/AMACR/ASAH2/PCK1/CYP3A4/ABHD2/CD36/ACOT11/PRKAB2/PGK1/GSTA1/PHYH/ACSS1/CYP2C18/ELOVL7/CD74/SLC27A2/UGT1A8/ACSF2/CES2/CYP1A1/ADTRP/ACOX2/GSTM2/CYP2J2/PCK2/ADH4/GPX4/LIPG/CYP2C9/PTGES2/MIF/IRS2/HPGD/CYP4F2/CYP2U1/ACADS/GSTM4/MGST3/ACSL5/THEM4/RGN/CBR1/SLC45A3/FAAH/CYP4F12/CYP2C19
4 MMP7/OTC/AKR1C1/ABCC2/SOD1/ABAT/BPHL/CYP2S1/NOS2/AADAC/ADD3/CYP3A7/GSTA2/CYP3A4/NFKBIZ/ACER2/SULT2A1/MAP2K6/CPS1/SLC6A4/GSTA1/CYP2C18/HNF4A/MAOB/ALDH2/UGT1A8/CALR/CES2/FMO5/CYP1A1/SLC28A2/GSTM2/GATA6/CYP2J2/ITGA3/CDH3/NAT2/CYP2C9/FBP1/CAT/SULT1A2/CYP4F2/C1orf115/CYP2U1/UGT2B7/HMGCS2/GSTM4/STAT1/CYP3A5/VAV3/TRPA1/CBR1/CYP4F12/ACE/CYP2C19
5                                                                                                                       AKR1C1/ABCC2/BPHL/CYP2S1/NOS2/AADAC/CYP3A7/GSTA2/CYP3A4/ACER2/SULT2A1/GSTA1/CYP2C18/HNF4A/ALDH2/UGT1A8/CES2/FMO5/CYP1A1/SLC28A2/GSTM2/CYP2J2/NAT2/CYP2C9/FBP1/SULT1A2/CYP4F2/CYP2U1/UGT2B7/GSTM4/CYP3A5/CBR1/CYP4F12/CYP2C19
6                                                   APOL1/AKR1B10/AKR1C1/DGAT1/TKFC/NUS1/CUBN/ASAH2/ACP3/SPHK2/PCK1/CYP3A7/CYP3A4/ACER2/SULT2A1/INPP5J/ACSS1/ACO2/CYP2C18/ALDH2/MOGAT2/FMO5/CYP1A1/NR0B2/PCK2/ADH4/SORD/APOA1/ALDH1A1/FDFT1/WNT4/CYP2C9/CAT/SULT1A2/CEL/HMGCS2/GK5/CYP3A5/GDE1/DEGS2/GDPD3/PTAFR/LPCAT3/PRKG2/EPHX2/DGKQ/IDH1/AKR7A2
  Count
1    31
2    51
3    53
4    55
5    34
6    48</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="co"># clusterProfiler dotplot</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dotplot</span>(compare.go, <span class="at">showCategory =</span> <span class="dv">50</span>,</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">x =</span> <span class="st">"Cluster"</span>, <span class="at">color =</span> <span class="st">"FoldEnrichment"</span>, <span class="at">size =</span> <span class="st">"geneRatio"</span>,</span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">font.size =</span> <span class="dv">5</span>, <span class="at">title =</span> <span class="st">"GO (Biological Process) over-representation | diff genes with qval &lt;= 0.001"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="03_clusterprofiler_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get data.frame of output</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>compare.go_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(compare.go)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>compare.go <span class="ot">&lt;-</span> <span class="fu">compareCluster</span>(<span class="at">geneCluster =</span> compare.list, <span class="at">fun =</span> <span class="st">"enrichGO"</span>, <span class="at">ont =</span> <span class="st">"CC"</span>, <span class="at">universe =</span> bg.genes<span class="sc">$</span>ENTREZID,</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>                             <span class="at">OrgDb =</span> <span class="st">"org.Hs.eg.db"</span>, <span class="at">pvalueCutoff =</span> <span class="fl">0.05</span>, <span class="at">pAdjustMethod =</span> <span class="st">"BH"</span>, <span class="at">qvalueCutoff =</span> <span class="fl">0.2</span>,</span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>                             <span class="at">readable =</span> <span class="cn">TRUE</span>)</span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(compare.go)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Cluster         ID                                      Description GeneRatio
1    il22 GO:0045177                              apical part of cell    67/853
2    il22 GO:0016324                           apical plasma membrane    60/853
3    il22 GO:0005903                                     brush border    28/853
4    il22 GO:0031526                            brush border membrane    19/853
5    il22 GO:0098862          cluster of actin-based cell projections    32/853
6    il22 GO:0140534 endoplasmic reticulum protein-containing complex    22/853
    BgRatio RichFactor FoldEnrichment    zScore       pvalue     p.adjust
1 478/18671  0.1401674       3.068071 10.021772 1.974263e-16 9.713374e-14
2 413/18671  0.1452785       3.179946  9.801925 1.495458e-15 3.678827e-13
3 109/18671  0.2568807       5.622767 10.590588 4.619331e-14 7.575703e-12
4  60/18671  0.3166667       6.931399 10.068508 9.793257e-12 1.139379e-09
5 173/18671  0.1849711       4.048764  8.814580 1.157906e-11 1.139379e-09
6 118/18671  0.1864407       4.080931  7.345698 1.634304e-08 1.340129e-06
        qvalue
1 8.395813e-14
2 3.179816e-13
3 6.548105e-12
4 9.848294e-10
5 9.848294e-10
6 1.158349e-06
                                                                                                                                                                                                                                                                                                                                                                                                                                                                      geneID
1 ANXA13/CYBRD1/DUOX2/DUOXA2/SLC6A20/MUC1/CEACAM5/PDZK1IP1/ABCC2/SLC26A3/ABCG2/CEACAM6/OSMR/SLC9A2/AQP1/MGAM/CEACAM1/RAB27B/SI/CUBN/SLC13A2/ENPP3/EPB41L4B/ACP3/PDZK1/SLC7A9/SLC9A3/SLC3A1/CD36/SLC9B2/SLC10A2/CLIC5/PROM2/SLC22A5/CPO/INSC/SLC5A12/SLC22A3/MYO1A/CEACAM20/SLC12A2/PLD1/SLC39A8/SLC23A1/DUOX1/STX3/SLC5A1/CA2/SLC36A1/CLCN5/AHCYL1/SLC39A14/CYP4F2/SLC26A2/SLC7A11/SLC34A2/SLC39A4/SLC17A5/ATP6V1B2/SLC11A2/RAB27A/PRKG2/ATP8B1/CDHR5/SLC2A2/CYP4F12/SLC28A1
2                                             ANXA13/CYBRD1/DUOX2/SLC6A20/MUC1/CEACAM5/PDZK1IP1/ABCC2/SLC26A3/ABCG2/CEACAM6/OSMR/SLC9A2/AQP1/MGAM/CEACAM1/RAB27B/SI/CUBN/SLC13A2/ENPP3/PDZK1/SLC7A9/SLC9A3/SLC3A1/CD36/SLC9B2/SLC10A2/CLIC5/PROM2/SLC22A5/CPO/SLC5A12/SLC22A3/MYO1A/CEACAM20/SLC12A2/PLD1/SLC39A8/SLC23A1/DUOX1/STX3/SLC5A1/SLC36A1/AHCYL1/SLC39A14/CYP4F2/SLC26A2/SLC34A2/SLC39A4/SLC17A5/ATP6V1B2/SLC11A2/RAB27A/PRKG2/ATP8B1/CDHR5/SLC2A2/CYP4F12/SLC28A1
3                                                                                                                                                                                                                                                                                          CYBRD1/SLC26A3/ABCG2/AQP1/SI/CUBN/PDZK1/SLC7A9/ADD3/SLC9A3/SLC3A1/CD36/VIL1/SLC22A5/SLC28A2/PLS1/TMPRSS15/MYO1A/SLC5A1/SLC7A11/SLC34A2/MME/SLC11A2/MYO1D/CDHR5/SLC2A2/SLC28A1/ACE
4                                                                                                                                                                                                                                                                                                                                            CYBRD1/SLC26A3/ABCG2/AQP1/CUBN/PDZK1/SLC7A9/SLC9A3/SLC3A1/CD36/SLC22A5/SLC28A2/SLC5A1/SLC7A11/SLC34A2/SLC11A2/CDHR5/SLC28A1/ACE
5                                                                                                                                                                                                                                                                 CYBRD1/SLC26A3/ABCG2/MYO3B/AQP1/SI/CUBN/PDZK1/SLC7A9/ADD3/SLC9A3/SLC3A1/CD36/CLIC5/VIL1/SLC22A5/SLC28A2/PLS1/TMPRSS15/MYO1A/SLC5A1/SLC7A11/SLC34A2/MME/SLC11A2/TRPA1/MYO1D/ATP8B1/CDHR5/SLC2A2/SLC28A1/ACE
6                                                                                                                                                                                                                                                                                                                                        HYOU1/DNAJB11/PPIB/B2M/PDIA3/SEC11C/OSTC/HM13/HSPA5/TAP1/SEC61G/CALR/PDIA6/SDF2L1/SEC61B/SRPRB/SPCS2/TAPBP/SPCS3/RPN1/DDOST/SEC61A1
  Count
1    67
2    60
3    28
4    19
5    32
6    22</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="co"># clusterProfiler dotplot</span></span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dotplot</span>(compare.go, <span class="at">showCategory =</span> <span class="dv">50</span>,</span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a>        <span class="at">x =</span> <span class="st">"Cluster"</span>, <span class="at">color =</span> <span class="st">"FoldEnrichment"</span>, <span class="at">size =</span> <span class="st">"geneRatio"</span>,</span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">font.size =</span> <span class="dv">5</span>, <span class="at">title =</span> <span class="st">"GO (Cellular Component) over-representation | diff genes with qval &lt;= 0.001"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="03_clusterprofiler_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get data.frame of output</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>compare.go_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(compare.go)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><strong>in progress</strong> Split upregulated and downregulated genes by computing fold changes</p>
</section>
</section>
<section id="gene-set-enrichment-analysis" class="level2">
<h2 class="anchored" data-anchor-id="gene-set-enrichment-analysis">Gene set enrichment analysis</h2>
<p><strong>in progress</strong></p>
<p><em>Recompute gene-level aggregated p-values for full gene list sorted by ranking metric (ascending qval i.e.&nbsp;high rank to low rank)</em></p>
</section>
</section>
<section id="notes" class="level1">
<h1>Notes</h1>
<ul>
<li>Technical replicates for samples <code>A2-7*</code>, <code>A2-11*</code> and <code>A2-12*</code> are in progress.</li>
<li>All samples have also been mapped and counted using STAR with <code>--quantMode GeneCounts</code> for differential expression with <strong>DESeq2</strong> - are the main observations replicated?</li>
</ul>
</section>
<section id="packages-and-session-info" class="level1">
<h1>Packages and session info</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>pkgs <span class="ot">&lt;-</span> <span class="fu">cite_packages</span>(<span class="at">output =</span> <span class="st">"table"</span>, <span class="at">out.dir =</span> <span class="st">"."</span>)</span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(pkgs)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="caption-top table table-sm table-striped small">
<thead>
<tr class="header">
<th style="text-align: left;">Package</th>
<th style="text-align: left;">Version</th>
<th style="text-align: left;">Citation</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">base</td>
<td style="text-align: left;">4.5.1</td>
<td style="text-align: left;"><span class="citation" data-cites="base">@base</span></td>
</tr>
<tr class="even">
<td style="text-align: left;">biomaRt</td>
<td style="text-align: left;">2.64.0</td>
<td style="text-align: left;"><span class="citation" data-cites="biomaRt">@biomaRt</span>….</td>
</tr>
<tr class="odd">
<td style="text-align: left;">clusterProfiler</td>
<td style="text-align: left;">4.16.0</td>
<td style="text-align: left;"><span class="citation" data-cites="cluster">@cluster</span>….</td>
</tr>
<tr class="even">
<td style="text-align: left;">dittoSeq</td>
<td style="text-align: left;">1.20.1</td>
<td style="text-align: left;"><span class="citation" data-cites="dittoSeq">@dittoSeq</span></td>
</tr>
<tr class="odd">
<td style="text-align: left;">enrichplot</td>
<td style="text-align: left;">1.28.4</td>
<td style="text-align: left;"><span class="citation" data-cites="enrichplot">@enrichplot</span></td>
</tr>
<tr class="even">
<td style="text-align: left;">gplots</td>
<td style="text-align: left;">3.2.0</td>
<td style="text-align: left;"><span class="citation" data-cites="gplots">@gplots</span></td>
</tr>
<tr class="odd">
<td style="text-align: left;">knitr</td>
<td style="text-align: left;">1.50</td>
<td style="text-align: left;"><span class="citation" data-cites="knitr20">@knitr20</span>….</td>
</tr>
<tr class="even">
<td style="text-align: left;">PCAtools</td>
<td style="text-align: left;">2.20.0</td>
<td style="text-align: left;"><span class="citation" data-cites="PCAtools">@PCAtools</span></td>
</tr>
<tr class="odd">
<td style="text-align: left;">rbioapi</td>
<td style="text-align: left;">0.8.3</td>
<td style="text-align: left;"><span class="citation" data-cites="rbioapi">@rbioapi</span></td>
</tr>
<tr class="even">
<td style="text-align: left;">RColorBrewer</td>
<td style="text-align: left;">1.1.3</td>
<td style="text-align: left;"><span class="citation" data-cites="RColorB">@RColorB</span>….</td>
</tr>
<tr class="odd">
<td style="text-align: left;">rmarkdown</td>
<td style="text-align: left;">2.30</td>
<td style="text-align: left;"><span class="citation" data-cites="rmarkdo">@rmarkdo</span>….</td>
</tr>
<tr class="even">
<td style="text-align: left;">sleuth</td>
<td style="text-align: left;">0.30.2</td>
<td style="text-align: left;"><span class="citation" data-cites="sleuth">@sleuth</span></td>
</tr>
<tr class="odd">
<td style="text-align: left;">SummarizedExperiment</td>
<td style="text-align: left;">1.38.1</td>
<td style="text-align: left;"><span class="citation" data-cites="Summari">@Summari</span>….</td>
</tr>
<tr class="even">
<td style="text-align: left;">tidyverse</td>
<td style="text-align: left;">2.0.0</td>
<td style="text-align: left;"><span class="citation" data-cites="tidyverse">@tidyverse</span></td>
</tr>
<tr class="odd">
<td style="text-align: left;">viridis</td>
<td style="text-align: left;">0.6.5</td>
<td style="text-align: left;"><span class="citation" data-cites="viridis">@viridis</span></td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.5.1 (2025-06-13)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 22.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0 
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_GB.UTF-8        LC_COLLATE=en_GB.UTF-8    
 [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8   
 [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       

time zone: Europe/London
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] grateful_0.3.0              rbioapi_0.8.3              
 [3] enrichplot_1.28.4           clusterProfiler_4.16.0     
 [5] sleuth_0.30.2               dittoSeq_1.20.1            
 [7] SummarizedExperiment_1.38.1 Biobase_2.68.0             
 [9] GenomicRanges_1.60.0        GenomeInfoDb_1.44.3        
[11] IRanges_2.42.0              S4Vectors_0.46.0           
[13] BiocGenerics_0.54.1         generics_0.1.4             
[15] MatrixGenerics_1.20.0       matrixStats_1.5.0          
[17] viridis_0.6.5               viridisLite_0.4.2          
[19] RColorBrewer_1.1-3          lubridate_1.9.4            
[21] forcats_1.0.1               stringr_1.6.0              
[23] dplyr_1.1.4                 purrr_1.2.0                
[25] readr_2.1.6                 tidyr_1.3.1                
[27] tibble_3.3.0                ggplot2_4.0.1              
[29] tidyverse_2.0.0            

loaded via a namespace (and not attached):
  [1] rstudioapi_0.17.1           jsonlite_2.0.0             
  [3] magrittr_2.0.4              ggtangle_0.0.8             
  [5] farver_2.1.2                rmarkdown_2.30             
  [7] fs_1.6.6                    vctrs_0.6.5                
  [9] memoise_2.0.1               ggtree_3.16.3              
 [11] htmltools_0.5.8.1           S4Arrays_1.8.1             
 [13] Rhdf5lib_1.30.0             SparseArray_1.8.1          
 [15] rhdf5_2.52.1                gridGraphics_0.5-1         
 [17] plyr_1.8.9                  cachem_1.1.0               
 [19] igraph_2.2.1                lifecycle_1.0.4            
 [21] pkgconfig_2.0.3             Matrix_1.7-4               
 [23] R6_2.6.1                    fastmap_1.2.0              
 [25] gson_0.1.0                  GenomeInfoDbData_1.2.14    
 [27] digest_0.6.38               aplot_0.2.9                
 [29] colorspace_2.1-2            patchwork_1.3.2            
 [31] AnnotationDbi_1.70.0        RSQLite_2.4.4              
 [33] org.Hs.eg.db_3.21.0         labeling_0.4.3             
 [35] timechange_0.3.0            httr_1.4.7                 
 [37] abind_1.4-8                 compiler_4.5.1             
 [39] bit64_4.6.0-1               withr_3.0.2                
 [41] S7_0.2.1                    BiocParallel_1.42.2        
 [43] DBI_1.2.3                   R.utils_2.13.0             
 [45] rappdirs_0.3.3              DelayedArray_0.34.1        
 [47] tools_4.5.1                 ape_5.8-1                  
 [49] R.oo_1.27.1                 glue_1.8.0                 
 [51] nlme_3.1-168                GOSemSim_2.34.0            
 [53] rhdf5filters_1.20.0         grid_4.5.1                 
 [55] reshape2_1.4.5              fgsea_1.34.2               
 [57] gtable_0.3.6                tzdb_0.5.0                 
 [59] R.methodsS3_1.8.2           data.table_1.17.8          
 [61] hms_1.1.4                   XVector_0.48.0             
 [63] ggrepel_0.9.6               pillar_1.11.1              
 [65] yulab.utils_0.2.1           splines_4.5.1              
 [67] treeio_1.32.0               lattice_0.22-5             
 [69] renv_1.1.5                  bit_4.6.0                  
 [71] tidyselect_1.2.1            GO.db_3.21.0               
 [73] SingleCellExperiment_1.30.1 Biostrings_2.76.0          
 [75] knitr_1.50                  gridExtra_2.3              
 [77] xfun_0.54                   pheatmap_1.0.13            
 [79] stringi_1.8.7               UCSC.utils_1.4.0           
 [81] lazyeval_0.2.2              ggfun_0.2.0                
 [83] yaml_2.3.10                 evaluate_1.0.5             
 [85] codetools_0.2-19            qvalue_2.40.0              
 [87] ggplotify_0.1.3             cli_3.6.5                  
 [89] Rcpp_1.1.0                  png_0.1-8                  
 [91] parallel_4.5.1              blob_1.2.4                 
 [93] DOSE_4.2.0                  tidytree_0.4.6             
 [95] scales_1.4.0                ggridges_0.5.7             
 [97] crayon_1.5.3                rlang_1.1.6                
 [99] cowplot_1.2.0               fastmatch_1.1-6            
[101] KEGGREST_1.48.1            </code></pre>
</div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>